<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Changelog • fishpond</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-news">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fishpond</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.7.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/swish.html">Swish: differential expression accounting for inferential uncertainty</a>
    </li>
    <li>
      <a href="../articles/allelic.html">SEESAW - Allelic expression analysis with Salmon and Swish</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      
    </div>

    <div class="section level2">
<h2 class="page-header" data-toc-text="2.7.1" id="fishpond-271">fishpond 2.7.1<a class="anchor" aria-label="anchor" href="#fishpond-271"></a></h2>
<ul><li>Corrected a bug in the two-group interaction (without pairing) functionality, when the groups were imbalanced, as identified by Samuel Chen. Fixes GitHub issue #35.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="2.5.4" id="fishpond-254">fishpond 2.5.4<a class="anchor" aria-label="anchor" href="#fishpond-254"></a></h2>
<ul><li>As CellRanger 7 includes both spliced and unspliced counts in their count matrix, we want to mimic this behavior by adding more pre-defined output formats in the loadFry function. We added “all” and “U+S+A” to include all counts in the count matrix. Moreover, now the “scRNA” output format has an “unspliced” field, which contains the unspliced count matrix.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="2.5.3" id="fishpond-253">fishpond 2.5.3<a class="anchor" aria-label="anchor" href="#fishpond-253"></a></h2>
<ul><li>Fix bug where salmonEC did not correct equivalence class names for going from 0-indexing to 1-indexing internally. In prior<br>
versions, to correctly link equivalence classes to gene names, users would have needed to manually add a value of 1 to the equivalence class names, which was erroneously not mentioned in the man files. After this bug fix, if the equivalence class identifier reads 1|2|8, then the equivalence class is immediatly compatible with the transcripts and their respective genes in rows 1, 2 and 8 of ‘tx2gene_matched’, without any further user intervention.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="2.5.1" id="fishpond-251">fishpond 2.5.1<a class="anchor" aria-label="anchor" href="#fishpond-251"></a></h2>
<ul><li>Fix plotAllelicGene() so that when samples have an allele with no expression at the gene level, it doesn’t throw an error trying to divide by 0.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="2.4.0" id="fishpond-240">fishpond 2.4.0<a class="anchor" aria-label="anchor" href="#fishpond-240"></a></h2>
<ul><li>For simple paired swish analysis, adding a <code>fast=1</code> method which uses a one-sample z-score on the paired LFCs (averaged over samples, then median over inferential replicates). The permutation is computed by changing the signs of the LFC matrix and recomputing z-scores. Testing on the vignette example, but using all the transcripts, the one-sample z-score method takes &lt; 20 seconds while the signed rank method takes &gt; 200 seconds (12x speedup), while they have a high rate of agreement on the detected set (30:1 in common vs discordant).</li>
<li>Removed the fast=0 methods that were previously implemented where ranks could optionally be recomputed for every permutation. This was much slower and didn’t have any appreciable benefit.</li>
<li>readEDS() has moved to the <code>eds</code> package, such that fishpond no longer requires <em>Rcpp</em> and C++ code compilation.</li>
<li>Fix bug identified by GitHub user @JosephLalli, where importAllelicCounts would find a1 and a2 strings internal to gene IDs, instead of at the suffix.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="2.3.22" id="fishpond-2322">fishpond 2.3.22<a class="anchor" aria-label="anchor" href="#fishpond-2322"></a></h2>
<ul><li>Fix bug identified by GitHub user @JosephLalli, where importAllelicCounts would find a1 and a2 strings internal to gene IDs, instead of at the suffix.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="2.3.14" id="fishpond-2314">fishpond 2.3.14<a class="anchor" aria-label="anchor" href="#fishpond-2314"></a></h2>
<ul><li>For simple paired swish analysis, adding a <code>fast=1</code> method which uses a one-sample z-score on the paired LFCs (averaged over samples, then median over inferential replicates). The permutation is computed by changing the signs of the LFC matrix and recomputing z-scores. Testing on the vignette example, but using all the transcripts, the one-sample z-score method takes &lt;20 seconds while the signed rank method takes &gt;200 seconds (12x speedup), while they have a high rate of agreement on the detected set ( 30:1 in common vs discordant).</li>
<li>Removed the fast=0 methods that were previously implemented where ranks could optionally be recomputed for every permutation. This was much slower and didn’t have any appreciable benefit.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="2.3.7" id="fishpond-237">fishpond 2.3.7<a class="anchor" aria-label="anchor" href="#fishpond-237"></a></h2>
<ul><li>readEDS() has moved to the <code>eds</code> package.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="2.2.0" id="fishpond-220">fishpond 2.2.0<a class="anchor" aria-label="anchor" href="#fishpond-220"></a></h2>
<ul><li>New vignette demonstrating allelic analysis at isoform, TSS, or gene-level. See more details below.</li>
<li>New import functions for equivalence class analysis of Salmon or alevin data, written by Jeroen Gilis. See salmonEC() and alevinEC() man pages.</li>
<li>New plotAllelicGene() and plotAllelicHeatmap() functions for plotting results from allelic expression analysis.</li>
<li>New makeTx2Tss() helper function for allelic analysis.</li>
<li>Now importFromAllelicCounts() can take a GRanges object as the <code>tx2gene</code> argument, so that ranges will be distributed to the rowRanges of the output SummarizedExperiment.</li>
<li>Adding <code>shiftX</code> argument to plotInfReps() for numeric x variable, to help with overplotting.</li>
<li>Re-organized package for new pkgdown homepage: <a href="https://mikelove.github.io/fishpond" class="uri">https://mikelove.github.io/fishpond</a>
</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="2.0.0" id="fishpond-200">fishpond 2.0.0<a class="anchor" aria-label="anchor" href="#fishpond-200"></a></h2>
<ul><li>New loadFry() function, written by Dongze He with contributions from Steve Lianoglou and Wes Wilson. loadFry() helps users to import and process alevin-fry quantification results. Can process spliced, unspliced and ambiguous counts separately and flexibly. Has specific output formats designed for downstream use with scVelo or velocity analysis. See ?loadFry for more details.</li>
<li>Adding correlation tests: Spearman or Pearson correlations of a numeric covariate with the log counts, or with the log fold changes across pairs. The Spearman correlation test with counts was already implemented in the original SAMseq method as response type = “Quantitative”. For new functionality see ‘cor’ argument in the ?swish man page.</li>
<li>Adding importAllelicCounts() to facilitate importing Salmon quantification data against a diploid transcriptome. Can import either as a ‘wide’ format or as ‘assays’. Leverages tximeta(). For gene-level summarization, importAllelicCounts() can create an appropriate tx2gene table with the necessary a1 and a2 suffices, and it will automatically set txOut=FALSE, see ?importAllelicCounts for more details.</li>
<li>Added a ‘q’ argument to plotInfReps to change the intervals when making point and line plots.</li>
<li>Switched the legend of plotInfReps so that reference levels will now be on the bottom, and non-reference (e.g. treatment) on top.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.99.18" id="fishpond-19918">fishpond 1.99.18<a class="anchor" aria-label="anchor" href="#fishpond-19918"></a></h2>
<ul><li>Added helper functionality to importAllelicCounts, so it will create an appropriate tx2gene table with the necessary a1 and a2 suffices, and it will automatically set txOut=FALSE.</li>
<li>Added a ‘q’ argument to plotInfReps to change the intervals when making point and line plots.</li>
<li>Switched the legend of plotInfReps so that reference levels will now be on the bottom, and non-reference (e.g. treatment) on top.</li>
<li>Added loadFry() to process alevin-fry quantification result. Can process spliced, unspliced and ambiguous counts separately and flexibly.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.99.15" id="fishpond-19915">fishpond 1.99.15<a class="anchor" aria-label="anchor" href="#fishpond-19915"></a></h2>
<ul><li>Adding correlation tests: Spearman or Pearson correlations of a numeric covariate with the log counts, or with the log fold changes across pairs. The Spearman correlation test with counts was already implemented in the original SAMseq method as response type = “Quantitative”. For new functionality see ‘cor’ argument in the ?swish man page.</li>
<li>Adding importAllelicCounts() to facilitate importing Salmon quantification data against a diploid transcriptome. Can import either as a ‘wide’ format or as ‘assays’. Leverages tximeta().</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.9.6" id="fishpond-196">fishpond 1.9.6<a class="anchor" aria-label="anchor" href="#fishpond-196"></a></h2>
<ul><li>Specifying ties.method in matrixStats::rowRanks.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.9.1" id="fishpond-191">fishpond 1.9.1<a class="anchor" aria-label="anchor" href="#fishpond-191"></a></h2>
<ul><li>Added importAllelicCounts() with options for importing Salmon quantification on diploid transcriptomes.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.8.0" id="fishpond-180">fishpond 1.8.0<a class="anchor" aria-label="anchor" href="#fishpond-180"></a></h2>
<ul><li>Added note in vignette about how to deal with estimated batch factors, e.g. from RUVSeq or SVA. Two strategies are outlined: either discretizing the estimate batch factors and performing stratified analysis, or regressing out the batch-associated variation using limma’s removeBatchEffect. Demonstation code is included.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.6.0" id="fishpond-160">fishpond 1.6.0<a class="anchor" aria-label="anchor" href="#fishpond-160"></a></h2>
<ul><li>Added makeInfReps() to create pseudo-inferential replicates via negative binomial simulation from mean and variance matrices. Note: the mean and the variance provide the <em>inferential</em> distribution per element of the count matrix. See preprint for details, doi: 10.1101/2020.07.06.189639.</li>
<li>Added splitSwish() and addStatsFromCSV(), which can be used to distribute running of Swish across a number of jobs managed by <code>Snakemake</code>. See vignette for description of a suggested workflow. For a large single-cell dataset with mean and variance summaries of inferential uncertainty, splitSwish() avoids generating the inferential replicate counts until the data has been split into smaller pieces and sent to different jobs, then only the necessary summary statistics are gathered and q-values computed by addStatsFromCSV().</li>
<li>plotInfReps() gains many new features to facilitate plotting of inferential count distributions for single cells, as quantified with alevin and imported with tximport. E.g. allow for numeric <code>x</code> argument plus grouping with <code>cov</code> for showing counts over pseudotime across groups of cells. Also added <code>applySF</code> argument which can be used to divide out a size factor, and the <code>reorder</code> argument which will re-order the samples/cells within groups by the count. plotInfReps() will draw boxplots with progressively thinner visual features as the number of cells grows to make the plots still legible.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.5.2" id="fishpond-152">fishpond 1.5.2<a class="anchor" aria-label="anchor" href="#fishpond-152"></a></h2>
<ul><li>First version of makeInfReps(), to create pseudo-infReps via negative binomial simulation from set of mean and variance matrices in the assays of the SummarizedExperiment.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.4.0" id="fishpond-140">fishpond 1.4.0<a class="anchor" aria-label="anchor" href="#fishpond-140"></a></h2>
<ul><li>Added isoformProportions(), which can be run after scaleInfReps() and optionally after filtering out transcripts using labelKeep(). Running swish() after isoformProportions() will produce differential transcript usage (DTU) results, instead of differential transcript expression (DTE) results. Example in vignette.</li>
<li>Default number of permutations increased from 30 to 100. It was observed that there was too much fluctuation in the DE called set for nperms=30 across different seeds, and setting to 100 helped to stabilize results across seeds, without increasing running time too much. For further reduced dependence on the seed, even higher values of nperms (e.g. 200, 300) can be used.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.8" id="fishpond-138">fishpond 1.3.8<a class="anchor" aria-label="anchor" href="#fishpond-138"></a></h2>
<ul><li>Added isoformProportions(), which can be run after scaleInfReps() and optionally after filtering out transcripts using labelKeep(). Running swish() after isoformProportions() will produce differential transcript usage (DTU) results, instead of differential transcript expression (DTE) results. Example in vignette.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.3.4" id="fishpond-134">fishpond 1.3.4<a class="anchor" aria-label="anchor" href="#fishpond-134"></a></h2>
<ul><li>Default number of permutations increased from 30 to 100. It was observed that there was too much fluctuation in the DE called set for nperms=30 across different seeds, and setting to 100 helped to stabilize results across seeds, without increasing running time too much. For further reduced dependence on the seed, even higher values of nperms (e.g. 200, 300) can be used.</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="1.2.0" id="fishpond-120">fishpond 1.2.0<a class="anchor" aria-label="anchor" href="#fishpond-120"></a></h2>
<ul><li>Switching to a faster version of Swish which only computes the ranks of the data once, and then re-uses this for the permutation distribution. This bypasses the addition of uniform noise per permutation and is 10x faster. Two designs which still require re-computation of ranks per permutation are the paired analysis and the general interaction analysis. Two-group, stratified two-group, and the paired interaction analysis now default to the new fast method, but the original, slower method can be used by setting fast=0 in the call to swish().</li>
<li>Adding Rcpp-based function readEDS() written by Avi Srivastava which imports the sparse counts stored in Alevin’s Efficient Data Storage (EDS) format.</li>
<li>Changed the vignette so that it (will) use a linkedTxome, as sometime the build would break if the Bioc build machine couldn’t access ftp.ebi.ac.uk.</li>
<li>Add ‘computeInfRV’ function. InfRV is not used in the Swish methods, only for visualization purposes in the Swish paper.</li>
<li>removed ‘samr’ from Imports, as it required source installation, moved to Suggests, for optional qvalue calculation</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.30" id="fishpond-09930">fishpond 0.99.30<a class="anchor" aria-label="anchor" href="#fishpond-09930"></a></h2>
<ul><li>added two interaction tests, described in ?swish</li>
<li>incorporate qvalue package for pvalue, locfdr and qvalue</li>
<li>added plotMASwish() to facilitate plotting</li>
<li>wilcoxP is removed, and the mean is used instead</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.99.0" id="fishpond-0990">fishpond 0.99.0<a class="anchor" aria-label="anchor" href="#fishpond-0990"></a></h2>
<ul><li>fishpond getting ready for submission to Bioc</li>
</ul></div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://www.linkedin.com/in/aqzhu91" class="external-link">Anqi Zhu</a>, <a href="https://mikelove.github.io" class="external-link">Michael Love</a>, <a href="https://k3yavi.github.io" class="external-link">Avi Srivastava</a>, <a href="https://combine-lab.github.io" class="external-link">Rob Patro</a>, <a href="https://sph.unc.edu/adv_profile/joseph-g-ibrahim-phd" class="external-link">Joseph Ibrahim</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

