<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Import allelic counts as a SummarizedExperiment — importAllelicCounts • fishpond</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Import allelic counts as a SummarizedExperiment — importAllelicCounts"><meta property="og:description" content="Read in Salmon quantification of allelic counts from a
diploid transcriptome. Assumes that diploid transcripts
are marked with the following suffix: an underscore and
a consistent symbol for each of the two alleles,
e.g. ENST123_M and ENST123_P,
or ENST123_alt and ENST123_ref, etc.
importAllelicCounts requires the tximeta package.
Further information in Details below."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fishpond</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.1.39</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/swish.html">Swish: differential expression accounting for inferential uncertainty</a>
    </li>
    <li>
      <a href="../articles/allelic.html">SEESAW - Allelic expression analysis with Salmon and Swish</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mikelove/fishpond/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Import allelic counts as a SummarizedExperiment</h1>
    <small class="dont-index">Source: <a href="https://github.com/mikelove/fishpond/blob/HEAD/R/helper-allelic.R" class="external-link"><code>R/helper-allelic.R</code></a></small>
    <div class="hidden name"><code>importAllelicCounts.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Read in Salmon quantification of allelic counts from a
diploid transcriptome. Assumes that diploid transcripts
are marked with the following suffix: an underscore and
a consistent symbol for each of the two alleles,
e.g. <code>ENST123_M</code> and <code>ENST123_P</code>,
or <code>ENST123_alt</code> and <code>ENST123_ref</code>, etc.
<code>importAllelicCounts</code> requires the tximeta package.
Further information in Details below.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">importAllelicCounts</span><span class="op">(</span>
  <span class="va">coldata</span>,
  <span class="va">a1</span>,
  <span class="va">a2</span>,
  format <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"wide"</span>, <span class="st">"assays"</span><span class="op">)</span>,
  tx2gene <span class="op">=</span> <span class="cn">NULL</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>coldata</dt>
<dd><p>a data.frame as used in <code>tximeta</code></p></dd>
<dt>a1</dt>
<dd><p>the symbol for the effect allele</p></dd>
<dt>a2</dt>
<dd><p>the symbol for the non-effect allele</p></dd>
<dt>format</dt>
<dd><p>either <code>"wide"</code> or <code>"assays"</code> for whether
to combine the allelic counts as columns (wide) or put the allelic
count information in different assay slots (assays).
For wide output, the data for the non-effect allele (a2) comes first,
then the effect allele (a1), e.g. <code>[a2 | a1]</code>. The <code>ref</code> level
of the factor variable <code>se$allele</code> will be <code>"a2"</code>
(so by default comparisons will be: a1 vs a2).
For assays output, all of the original matrices are renamed with a prefix,
either <code>a1-</code> or <code>a2-</code>.</p></dd>
<dt>tx2gene</dt>
<dd><p>optional, a data.frame with first column indicating
transcripts, second column indicating genes (or any other transcript
grouping). Alternatively, this can be a GRanges object with
required columns <code>tx_id</code>, and <code>group_id</code>
(see <code>makeTx2Tss</code>). For more information on this argument,
see Details.</p></dd>
<dt>...</dt>
<dd><p>any arguments to pass to tximeta</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>a SummarizedExperiment, with allele counts (and other data)
combined into a wide matrix <code>[a2 | a1]</code>, or as assays (a1, then a2).
The original strings associated with a1 and a2 are stored in the
metadata of the object, in the <code>alleles</code> list element.
Note the reference level of <code>se$allele</code> will be <code>"a2"</code>, 
such that comparisons by default will be a1 vs a2
(effect vs non-effect).</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p><strong>Requirements</strong> - There must be exactly two alleles for each
transcript, and the <code>--keep-duplicates</code> option should be used
in Salmon indexing to avoid removal of transcripts with identical
sequence. The output object has half the number of transcripts,
with the two alleles either stored in a <code>"wide"</code> object, or as
re-named <code>"assays"</code>. Note carefully that the symbol provided
to <code>a1</code> is used as the effect allele, and <code>a2</code> is used as
the non-effect allele (see the <code>format</code> argument description
and Value description below).</p>
<p><strong>tx2gene</strong> - The two columns should include the <code>a1</code> and
<code>a2</code> suffix for the transcripts and genes/groups, or those
will be added internally, if it is detected that the first
transcript does not have these suffices. For example if
<code>_alt</code> or <code>_ref</code>, or <code>_M</code> or <code>_P</code> (as indicated
by the <code>a1</code> and <code>a2</code> arguments) are not present in the
table, the table rows will be duplicated with those suffices added
on behalf of the user. If <code>tx2gene</code> is not provided, the
output object will be transcript-level. Do not attempt to set the
<code>txOut</code> argument, it will conflict with internal calls to
downstream functions. If the a1/a2 suffices are not at the end of
the transcript name in the quantification files,
e.g. <code>ENST123_M|&lt;metadata&gt;</code>, then <code>ignoreAfterBar=TRUE</code>
can be used to match regardless of the string following <code>|</code> in
the quantification files.</p>
<p><code>skipMeta=TRUE</code> is used, as it is assumed the diploid
transcriptome does not match any reference transcript
collection. This may change in future iterations of the function,
depending on developments in upstream annotations and software.</p>
<p>If <code>tx2gene</code> is a GRanges object, the rowRanges of the output
will be the reduced ranges of the grouped input ranges, with
<code>tx_id</code> collapsed into a CharacterList. Other metadata columns
are not manipulated, just the metadata for the first range is
returned.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://www.linkedin.com/in/aqzhu91" class="external-link">Anqi Zhu</a>, <a href="https://mikelove.github.io" class="external-link">Michael Love</a>, <a href="https://k3yavi.github.io" class="external-link">Avi Srivastava</a>, <a href="https://combine-lab.github.io" class="external-link">Rob Patro</a>, <a href="https://sph.unc.edu/adv_profile/joseph-g-ibrahim-phd" class="external-link">Joseph Ibrahim</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

