<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Plot allelic counts in a gene context using Gviz — plotAllelicGene • fishpond</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Plot allelic counts in a gene context using Gviz — plotAllelicGene"><meta property="og:description" content="Plot allelic data (allelic proportions, isoform propostions)
in a gene context leveraging the Gviz package. See the allelic
vignette for example usage. TPM and count filters are used by
default to clean up the plot of features with minimal signal;
note that the isoform proportion displayed at the bottom of the
plot is among the features that pass the filtering steps.
If the function is not responding, it is likely due to issues
connecting to UCSC servers to draw the ideogram, in this case
set ideogram=FALSE."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fishpond</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.3.12</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/swish.html">Swish: differential expression accounting for inferential uncertainty</a>
    </li>
    <li>
      <a href="../articles/allelic.html">SEESAW - Allelic expression analysis with Salmon and Swish</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mikelove/fishpond/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot allelic counts in a gene context using Gviz</h1>
    <small class="dont-index">Source: <a href="https://github.com/mikelove/fishpond/blob/HEAD/R/helper-allelic.R" class="external-link"><code>R/helper-allelic.R</code></a></small>
    <div class="hidden name"><code>plotAllelicGene.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Plot allelic data (allelic proportions, isoform propostions)
in a gene context leveraging the Gviz package. See the allelic
vignette for example usage. TPM and count filters are used by
default to clean up the plot of features with minimal signal;
note that the isoform proportion displayed at the bottom of the
plot is among the features that pass the filtering steps.
If the function is not responding, it is likely due to issues
connecting to UCSC servers to draw the ideogram, in this case
set <code>ideogram=FALSE</code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plotAllelicGene</span><span class="op">(</span></span>
<span>  <span class="va">y</span>,</span>
<span>  <span class="va">gene</span>,</span>
<span>  <span class="va">db</span>,</span>
<span>  region <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  symbol <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  genome <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  tpmFilter <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  isoPropFilter <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  countFilter <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  pc <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  transcriptAnnotation <span class="op">=</span> <span class="st">"symbol"</span>,</span>
<span>  labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>a2 <span class="op">=</span> <span class="st">"a2"</span>, a1 <span class="op">=</span> <span class="st">"a1"</span><span class="op">)</span>,</span>
<span>  qvalue <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  log2FC <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ideogram <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  cov <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  covFacetIsoform <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  allelicCol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dodgerblue"</span>, <span class="st">"goldenrod1"</span><span class="op">)</span>,</span>
<span>  isoformCol <span class="op">=</span> <span class="st">"firebrick"</span>,</span>
<span>  statCol <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  gridCol <span class="op">=</span> <span class="st">"grey80"</span>,</span>
<span>  baselineCol <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  titleCol <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  titleAxisCol <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  titleBgCol <span class="op">=</span> <span class="st">"white"</span>,</span>
<span>  geneBorderCol <span class="op">=</span> <span class="st">"darkblue"</span>,</span>
<span>  geneFillCol <span class="op">=</span> <span class="st">"darkblue"</span>,</span>
<span>  genomeAxisCol <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  innerFontCol <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>y</dt>
<dd><p>a SummarizedExperiment (see <code>swish</code>)</p></dd>


<dt>gene</dt>
<dd><p>the name of the gene of interest, requires
a column <code>gene_id</code> in the metadata columns of the
rowRanges of y</p></dd>


<dt>db</dt>
<dd><p>either a TxDb or EnsDb object to use for the gene model</p></dd>


<dt>region</dt>
<dd><p>GRanges, the region to be displayed in the Gviz plot.
if not specified, will be set according to the gene plus 20<!-- % -->
of the total gene extent on either side</p></dd>


<dt>symbol</dt>
<dd><p>alternative to <code>gene</code>, to specify
the gene of interest according to a column <code>symbol</code>
in the metadata columns of the rowRanges of y</p></dd>


<dt>genome</dt>
<dd><p>UCSC genome code (e.g. <code>"hg38"</code>,
if not specified it will use the <code><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">GenomeInfoDb::genome()</a></code>
of the rowRanges of <code>y</code></p></dd>


<dt>tpmFilter</dt>
<dd><p>minimum TPM value (mean over samples) to keep a feature</p></dd>


<dt>isoPropFilter</dt>
<dd><p>minimum percent of isoform proportion to keep a feature</p></dd>


<dt>countFilter</dt>
<dd><p>minimum count value (mean over samples) to keep a feature</p></dd>


<dt>pc</dt>
<dd><p>pseudocount to avoid dividing by zero in allelic proportion calculation</p></dd>


<dt>transcriptAnnotation</dt>
<dd><p>argument passed to Gviz::GeneRegionTrack
(<code>"symbol"</code>, <code>"gene"</code>, <code>"transcript"</code>, etc.)</p></dd>


<dt>labels</dt>
<dd><p>list, labels for a2 (non-effect) and a1 (effect) alleles</p></dd>


<dt>qvalue</dt>
<dd><p>logical, whether to inclue qvalue track</p></dd>


<dt>log2FC</dt>
<dd><p>logical, whether to include log2FC track</p></dd>


<dt>ideogram</dt>
<dd><p>logical, whether to include ideogram track</p></dd>


<dt>cov</dt>
<dd><p>character specifying a factor or integer variable to use
to facet the allelic proportion plots, should be a column in
<code>colData(y)</code></p></dd>


<dt>covFacetIsoform</dt>
<dd><p>logical, if <code>cov</code> is provided,
should it also be used to facet the isoform proportion track,
in addition to the allelic proportion track</p></dd>


<dt>allelicCol</dt>
<dd><p>the colors of the points and lines for allelic proportion</p></dd>


<dt>isoformCol</dt>
<dd><p>the colors of the points and lines for isoform proportion</p></dd>


<dt>statCol</dt>
<dd><p>the color of the lollipops for q-value and log2FC</p></dd>


<dt>gridCol</dt>
<dd><p>the color of the grid in the data tracks</p></dd>


<dt>baselineCol</dt>
<dd><p>the color of the horizontal baseline for q-value and lo2gFC</p></dd>


<dt>titleCol</dt>
<dd><p>font color of the side titles (track labels)</p></dd>


<dt>titleAxisCol</dt>
<dd><p>axis color of the side titles (track labels)</p></dd>


<dt>titleBgCol</dt>
<dd><p>background color of the side titles (track labels)</p></dd>


<dt>geneBorderCol</dt>
<dd><p>the color of the borders and font in gene region track</p></dd>


<dt>geneFillCol</dt>
<dd><p>the color of the fill in the gene region track</p></dd>


<dt>genomeAxisCol</dt>
<dd><p>line color of the genome axis track</p></dd>


<dt>innerFontCol</dt>
<dd><p>font color of genome axis track, ideogram, and
allelic proportion legend</p></dd>


<dt>...</dt>
<dd><p>additional arguments passed to <code><a href="https://rdrr.io/pkg/Gviz/man/plotTracks.html" class="external-link">Gviz::plotTracks()</a></code></p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>nothing, a plot is displayed</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://www.linkedin.com/in/aqzhu91" class="external-link">Anqi Zhu</a>, <a href="https://mikelove.github.io" class="external-link">Michael Love</a>, <a href="https://k3yavi.github.io" class="external-link">Avi Srivastava</a>, <a href="https://combine-lab.github.io" class="external-link">Rob Patro</a>, <a href="https://sph.unc.edu/adv_profile/joseph-g-ibrahim-phd" class="external-link">Joseph Ibrahim</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

