<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Construct a sparse matrix of transcript compatibility counts from salmon 
output — salmonEC • fishpond</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Construct a sparse matrix of transcript compatibility counts from salmon 
output — salmonEC"><meta property="og:description" content="Constructs a count matrix with equivalence class identifiers
in the rows. The count matrix is generated from one or multiple 
`eq_classes.txt` files that have been created by running salmon with the 
--dumpEq flag. Salmon - https://doi.org/10.1038/nmeth.4197"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fishpond</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/swish.html">Swish: differential expression accounting for inferential uncertainty</a>
    </li>
    <li>
      <a href="../articles/allelic.html">SEESAW - Allelic expression analysis with Salmon and Swish</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/mikelove/fishpond/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Construct a sparse matrix of transcript compatibility counts from salmon 
output</h1>
    <small class="dont-index">Source: <a href="https://github.com/mikelove/fishpond/blob/HEAD/R/salmon-EC.R" class="external-link"><code>R/salmon-EC.R</code></a></small>
    <div class="hidden name"><code>salmonEC.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Constructs a count matrix with equivalence class identifiers
in the rows. The count matrix is generated from one or multiple 
`eq_classes.txt` files that have been created by running salmon with the 
--dumpEq flag. Salmon - <a href="https://doi.org/10.1038/nmeth.4197" class="external-link">https://doi.org/10.1038/nmeth.4197</a></p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">salmonEC</span><span class="op">(</span></span>
<span>  <span class="va">paths</span>,</span>
<span>  <span class="va">tx2gene</span>,</span>
<span>  multigene <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ignoreTxVersion <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ignoreAfterBar <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>paths</dt>
<dd><p>`Charachter` or `character vector`, path specifying the 
location of the `eq_classes.txt` files generated with salmon.</p></dd>


<dt>tx2gene</dt>
<dd><p>A `dataframe` linking transcript identifiers to their 
corresponding gene identifiers. Transcript identifiers must be in a column
`isoform_id`. Corresponding gene identifiers must be in a column `gene_id`.</p></dd>


<dt>multigene</dt>
<dd><p>`Logical`, should equivalence classes that are compatible 
with multiple genes be retained? Default is `FALSE`, removing such ambiguous
equivalence classes.</p></dd>


<dt>ignoreTxVersion</dt>
<dd><p>logical, whether to split the isoform id on the '.' 
character to remove version information to facilitate matching with the
isoform id in `tx2gene` (default FALSE).</p></dd>


<dt>ignoreAfterBar</dt>
<dd><p>logical, whether to split the isoform id on the '|' 
character to facilitate matching with the isoform id in `tx2gene` 
(default FALSE).</p></dd>


<dt>quiet</dt>
<dd><p>`Logical`, set `TRUE` to avoid displaying messages.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A list with two elements. The first element `counts` is a sparse 
count matrix with equivalence class identifiers in the rows. If multiple 
paths are specified, the columns are in the same order as the paths. The 
second element `tx2gene_matched` allows for linking those identifiers to
their respective transcripts and genes.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    

<p>The resulting count matrix uses equivalence class identifiers as rownames.
These can be linked to respective transcripts and genes using the 
`tx2gene_matched` element of the output. Specifically, if the equivalence 
class identifier reads 1|2|8, then the equivalence class is compatible with
the transcripts and their respective genes in rows 1, 2 and 8 of 
`tx2gene_matched`.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Jeroen Gilis</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://www.linkedin.com/in/aqzhu91" class="external-link">Anqi Zhu</a>, <a href="https://mikelove.github.io" class="external-link">Michael Love</a>, <a href="https://k3yavi.github.io" class="external-link">Avi Srivastava</a>, <a href="https://combine-lab.github.io" class="external-link">Rob Patro</a>, <a href="https://sph.unc.edu/adv_profile/joseph-g-ibrahim-phd" class="external-link">Joseph Ibrahim</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

