<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Swish: differential expression accounting for inferential uncertainty • fishpond</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Swish: differential expression accounting for inferential uncertainty">
<meta property="og:description" content="fishpond">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">fishpond</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.7.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/swish.html">Swish: differential expression accounting for inferential uncertainty</a>
    </li>
    <li>
      <a href="../articles/allelic.html">SEESAW - Allelic expression analysis with Salmon and Swish</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mikelove/fishpond/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Swish: differential expression accounting for
inferential uncertainty</h1>
            
            <h4 data-toc-skip class="date">06/19/2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/mikelove/fishpond/blob/HEAD/vignettes/swish.Rmd" class="external-link"><code>vignettes/swish.Rmd</code></a></small>
      <div class="hidden name"><code>swish.Rmd</code></div>

    </div>

    
    
<!-- run this document with rmarkdown::render("swish.Rmd") -->
<div class="section level2">
<h2 id="the-swish-method">The Swish method<a class="anchor" aria-label="anchor" href="#the-swish-method"></a>
</h2>
<p>The <em>Swish</em> method for differential expression analysis of
bulk or single-cell RNA-seq data using inferential replicate counts is
described in the following reference: <span class="citation">Zhu et al.
(2019)</span> <a href="https://doi.org/10.1093/nar/gkz622" class="external-link">doi:
10.1093/nar/gkz622</a>.</p>
<p>We note that <em>Swish</em> extends and builds on another method,
<em>SAMseq</em> <span class="citation">(Li and Tibshirani 2011)</span>,
implemented in the <em>samr</em> package, by taking into account
inferential uncertainty, and allowing to control for batch effects and
matched samples. Additionally, <em>Swish</em> has methods for testing
changes in effect size across secondary covariates, which we refer to as
“interactions”. <em>Swish</em> also implements correlation tests as in
the original <em>SAMseq</em> package. <code>swish</code> calls functions
from the <em>qvalue</em> <span class="citation">(Storey and Tibshirani
2003)</span> or <em>samr</em> package for calculation of local FDR and
q-value. This vignette gives an example of differential analysis of
matched samples, and an interaction test for matched samples, to see if
a condition effect changes in magnitude across two groups of
samples.</p>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>The following lines of code will perform a basic transcript-level
<code>swish</code> two group analysis of bulk RNA-seq. For more details,
read on. There is a special section below for two-group analysis of <a href="#alevin_scrna-seq">scRNA-seq</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 'coldata.csv': sample information table</span></span>
<span><span class="va">coldata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"coldata.csv"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/tximeta" class="external-link">tximeta</a></span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tximeta/man/tximeta.html" class="external-link">tximeta</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span> <span class="co"># reads in counts and inf reps</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mikelove.github.io/fishpond">fishpond</a></span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scaleInfReps.html">scaleInfReps</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span> <span class="co"># scales counts</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/labelKeep.html">labelKeep</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span> <span class="co"># labels features to keep</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/swish.html">swish</a></span><span class="op">(</span><span class="va">y</span>, x<span class="op">=</span><span class="st">"condition"</span><span class="op">)</span> <span class="co"># simplest Swish case</span></span></code></pre></div>
<p><strong>Note:</strong> Inferential replicates, either from Gibbs
sampling or bootstrapping of reads, are required for the <em>swish</em>
method. When running Salmon, you can set
<code>--numGibbsSamples 30</code> to generate Gibbs samples, or
<code>--numBootstraps 30</code> to generate bootstrap samples (we
typically recommend 20-30 inferential replicates).</p>
<p>The results can be found in <code>mcols(y)</code>. For example, one
can calculate the number of genes passing a 5% FDR threshold:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">qvalue</span> <span class="op">&lt;</span> <span class="fl">.05</span><span class="op">)</span></span></code></pre></div>
<p>One can at any point remove the genes that didn’t pass the expression
filter with the following line of code (can be run before or after
<code>swish</code>). These genes are ignored by <code>swish</code>, and
so will have <code>NA</code> in the results columns in
<code>mcols(y)</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">keep</span>,<span class="op">]</span></span></code></pre></div>
<p>A gene-level analysis looks identical to a transcript-level analysis,
only the input data changes. Examples follow.</p>
<p>Lastly, what is the structure of the output of <code>tximeta</code>
<span class="citation">(Love et al. 2020)</span>, which is used in
<code>swish</code>? See the section below, <em>Structure of tximeta
output / swish input</em>.</p>
<div class="section level3">
<h3 id="macrophage-stimulation-experiment">Macrophage stimulation experiment<a class="anchor" aria-label="anchor" href="#macrophage-stimulation-experiment"></a>
</h3>
<p>We begin the <em>fishpond</em> vignette by loading data from a
Bioconductor Experiment Data package, <em>macrophage</em>. The package
contains RNA-seq quantification from 24 RNA-seq samples, which are a
subset of the RNA-seq samples generated and analyzed by <span class="citation">Alasoo et al. (2018)</span> - <a href="https://doi.org/10.1038/s41588-018-0046-7" class="external-link">doi:
10.1038/s41588-018-0046-7</a>.</p>
<p>The experiment involved treatment of macrophage cell lines from a
number of human donors with IFN gamma, <em>Salmonella</em> infection, or
both treatments combined. In the beginning of this vignette, we will
focus on comparing the IFN gamma stimulated cell lines with the control
cell lines, accounting for the paired nature of the data (cells from the
same donor). Later in the vignette we will analyze differences in the
<em>Salmonella</em> infection response by IFN gamma treatment status –
whether the cells are primed for immune response.</p>
<p>We load the package, and point to the <code>extdata</code> directory.
For a typical analysis, the user would just point <code>dir</code> to
the location on the machine or cluster where the transcript
quantifications are stored (e.g. the <code>quant.sf</code> files).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">macrophage</span><span class="op">)</span></span>
<span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"macrophage"</span><span class="op">)</span></span></code></pre></div>
<p>The data was quantified using <em>Salmon</em> <span class="citation">(Patro et al. 2017)</span> 0.12.0 against the Gencode
v29 human reference transcripts <span class="citation">(Frankish,
GENCODE-consoritum, and Flicek 2018)</span>. For more details and all
code used for quantification, refer to the <a href="https://bioconductor.org/packages/macrophage" class="external-link">macrophage</a>
package vignette.</p>
<p>Importantly, <code>--numGibbsSamples 20</code> was used to generate
20 inferential replicates with <em>Salmon</em>’s Gibbs sampling
procedure. We also recommend to use <code>--gcBias</code> when running
<em>Salmon</em> to protect against common sample-specific biases present
in RNA-seq data.</p>
</div>
</div>
<div class="section level2">
<h2 id="data-import">Data import<a class="anchor" aria-label="anchor" href="#data-import"></a>
</h2>
<div class="section level3">
<h3 id="read-in-the-column-data-from-csv">Read in the column data from CSV<a class="anchor" aria-label="anchor" href="#read-in-the-column-data-from-csv"></a>
</h3>
<p>We start by reading in a CSV with the <em>column data</em>, that is,
information about the samples, which are represented as columns of the
<em>SummarizedExperiment</em> object we will construct containing the
counts of reads per gene or transcript.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coldata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"coldata.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            names sample_id line_id replicate condition_name macrophage_harvest</span></span>
<span><span class="co">## 1 SAMEA103885102    diku_A  diku_1         1          naive          11/6/2015</span></span>
<span><span class="co">## 2 SAMEA103885347    diku_B  diku_1         1           IFNg          11/6/2015</span></span>
<span><span class="co">## 3 SAMEA103885043    diku_C  diku_1         1         SL1344          11/6/2015</span></span>
<span><span class="co">## 4 SAMEA103885392    diku_D  diku_1         1    IFNg_SL1344          11/6/2015</span></span>
<span><span class="co">## 5 SAMEA103885182    eiwy_A  eiwy_1         1          naive         11/25/2015</span></span>
<span><span class="co">## 6 SAMEA103885136    eiwy_B  eiwy_1         1           IFNg         11/25/2015</span></span>
<span><span class="co">##   salmonella_date ng_ul_mean rna_extraction rna_submit        library_pool</span></span>
<span><span class="co">## 1      11/13/2015   293.9625     11/30/2015  12/9/2015 3226_RNAauto-091215</span></span>
<span><span class="co">## 2      11/13/2015   293.9625     11/30/2015  12/9/2015 3226_RNAauto-091215</span></span>
<span><span class="co">## 3      11/13/2015   293.9625     11/30/2015  12/9/2015 3226_RNAauto-091215</span></span>
<span><span class="co">## 4      11/13/2015   293.9625     11/30/2015  12/9/2015 3226_RNAauto-091215</span></span>
<span><span class="co">## 5       12/2/2015   193.5450      12/3/2015  12/9/2015 3226_RNAauto-091215</span></span>
<span><span class="co">## 6       12/2/2015   193.5450      12/3/2015  12/9/2015 3226_RNAauto-091215</span></span>
<span><span class="co">##   chemistry rna_auto</span></span>
<span><span class="co">## 1   V4_auto        1</span></span>
<span><span class="co">## 2   V4_auto        1</span></span>
<span><span class="co">## 3   V4_auto        1</span></span>
<span><span class="co">## 4   V4_auto        1</span></span>
<span><span class="co">## 5   V4_auto        1</span></span>
<span><span class="co">## 6   V4_auto        1</span></span></code></pre>
<p>We will subset to certain columns of interest, and re-name them for
later.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coldata</span> <span class="op">&lt;-</span> <span class="va">coldata</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">5</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"names"</span>,<span class="st">"id"</span>,<span class="st">"line"</span>,<span class="st">"condition"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="add-a-column-pointing-to-your-files">Add a column pointing to your files<a class="anchor" aria-label="anchor" href="#add-a-column-pointing-to-your-files"></a>
</h3>
<p><code>coldata</code> needs to have a column <code>files</code> which
specifies the path to the quantification files. In this case, we’ve
gzipped the quantification files, so we point to the
<code>quant.sf.gz</code> file. We make sure that all the files exist in
the location we specified.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coldata</span><span class="op">$</span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"quants"</span>, <span class="va">coldata</span><span class="op">$</span><span class="va">names</span>, <span class="st">"quant.sf.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">$</span><span class="va">files</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="read-in-quants-with-tximeta">Read in quants with tximeta<a class="anchor" aria-label="anchor" href="#read-in-quants-with-tximeta"></a>
</h3>
<p>We will read in quantification data for some of the samples. First we
load the <em>SummarizedExperiment</em> package. We will store out data
and the output of the statistical method in a
<em>SummarizedExperiment</em> object. We use the <em>tximeta</em> <span class="citation">(Love et al. 2020)</span> package to read in the
data:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The <em>macrophage</em> package contains a number of samples under
different conditions. In this vignette, we will demonstrate a two group
comparison, and so we first subset our <code>coldata</code> to the
<code>"naive"</code> and <code>"IFNg"</code> groups. Log fold changes
will be made comparing <code>IFNg</code> to <code>naive</code> (the
reference level).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coldata</span> <span class="op">&lt;-</span> <span class="va">coldata</span><span class="op">[</span><span class="va">coldata</span><span class="op">$</span><span class="va">condition</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"naive"</span>,<span class="st">"IFNg"</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">coldata</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">$</span><span class="va">condition</span>, </span>
<span>                            levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"naive"</span>,<span class="st">"IFNg"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We load in the quantification data with <code>tximeta</code>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/tximeta" class="external-link">tximeta</a></span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tximeta/man/tximeta.html" class="external-link">tximeta</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span></span></code></pre></div>
<p>We can see that all the assays have been loaded:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "counts"    "abundance" "length"    "infRep1"   "infRep2"   "infRep3"  </span></span>
<span><span class="co">##  [7] "infRep4"   "infRep5"   "infRep6"   "infRep7"   "infRep8"   "infRep9"  </span></span>
<span><span class="co">## [13] "infRep10"  "infRep11"  "infRep12"  "infRep13"  "infRep14"  "infRep15" </span></span>
<span><span class="co">## [19] "infRep16"  "infRep17"  "infRep18"  "infRep19"  "infRep20"</span></span></code></pre>
<p><code>tximeta</code> loads transcript-level data, although it can
later be summarized to the gene levels:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "ENST00000456328.2" "ENST00000450305.2" "ENST00000488147.1"</span></span>
<span><span class="co">## [4] "ENST00000619216.1" "ENST00000473358.1" "ENST00000469289.1"</span></span></code></pre>
<p>We will rename our <em>SummarizedExperiment</em> <code>y</code> for
the statistical analysis. For speed of the vignette, we subset to the
transcripts on chromosome 4.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">se</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="fu">seqnames</span><span class="op">(</span><span class="va">y</span><span class="op">)</span> <span class="op">==</span> <span class="st">"chr4"</span>,<span class="op">]</span></span></code></pre></div>
<p><strong>Note on factor levels</strong>: The <code>swish</code>
function compares expression level across factors such that log2 fold
changes are reported as the non-reference level over the reference
level. By default, R will choose a <em>reference level</em> for factors
based on alphabetical order, unless <code>levels</code> are explicitly
set. It is recommended to set the factors levels, as in the above code
chunk, with the reference level coming first in the character vector, so
that log2 fold changes correspond to the comparison of interest.</p>
</div>
</div>
<div class="section level2">
<h2 id="differential-transcript-expression">Differential transcript expression<a class="anchor" aria-label="anchor" href="#differential-transcript-expression"></a>
</h2>
<div class="section level3">
<h3 id="running-swish-at-the-transcript-level">Running Swish at the transcript level<a class="anchor" aria-label="anchor" href="#running-swish-at-the-transcript-level"></a>
</h3>
<p>Running <code>swish</code> has three steps: scaling the inferential
replicates, labeling the rows with sufficient counts for running
differential expression, and then calculating the statistics. As
<code>swish</code> makes use of pseudo-random number generation in
breaking ties and in calculating permutations, to obtain identical
results, one needs to set a random seed before running
<code><a href="../reference/swish.html">swish()</a></code>, as we do below.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mikelove.github.io/fishpond">fishpond</a></span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scaleInfReps.html">scaleInfReps</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/labelKeep.html">labelKeep</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">keep</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/swish.html">swish</a></span><span class="op">(</span><span class="va">y</span>, x<span class="op">=</span><span class="st">"condition"</span>, pair<span class="op">=</span><span class="st">"line"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Note:</strong> the simple paired test is the slowest of all
the other designs provided for Swish (see <a href="#further-details">further details</a>), because it requires
recomputing the signed ranks at each permutation. Other designs avoid
recomputing ranks per permutation, or don’t use ranks for the test
statistic. You can set <code>fast=1</code> for simple paired designs,
which implements a one-sample z-score as the test statistic, and
permutations to provide p-values and q-values. These permutations are
much faster than those that use the signed rank test, for example with
one core, the signed rank test is 12x slower than the one-sample z-score
method when run across all transcripts, and they nevertheless detect
nearly the same sets.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y_fast</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/swish.html">swish</a></span><span class="op">(</span><span class="va">y</span>, x<span class="op">=</span><span class="st">"condition"</span>, pair<span class="op">=</span><span class="st">"line"</span>, fast<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span>original<span class="op">=</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">qvalue</span> <span class="op">&lt;</span> <span class="fl">.1</span>,</span>
<span>      fast<span class="op">=</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y_fast</span><span class="op">)</span><span class="op">$</span><span class="va">qvalue</span> <span class="op">&lt;</span> <span class="fl">.1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         fast</span></span>
<span><span class="co">## original FALSE TRUE</span></span>
<span><span class="co">##    FALSE  1615   40</span></span>
<span><span class="co">##    TRUE     41  553</span></span></code></pre>
<p>The default number of permutations for computing p-values is
<code>nperms=100</code>. However, for paired datasets as this one, you
may have fewer maximum permutations. In this case, there are 64 possible
ways to switch the condition labels for six pairs of samples. We can set
the <code>nperms</code> manually to 64, or if we had just used the
default value, <code>swish</code> would set <code>nperms</code> to the
maximum value possible and notify the user that it had done so.</p>
<p>A note about <code>labelKeep</code>: by default we keep features with
<code>minN=3</code> samples with a minimal count of 10. For scRNA-seq
data with de-duplicated UMI counts, we recommend to lower the count,
e.g. a count of 3, across a higher number of <code>minN</code> cells,
depending on the number of cells being compared. You can also set
<code>x="condition"</code> when running <code>labelKeep</code> which
will use the condition variable to set <code>minN</code>.</p>
<p>The results are stored in <code>mcols(y)</code>. We will show below
how to pull out the top up- and down-regulated transcripts.</p>
<p>We can see how many transcripts are in a 5% FDR set:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">qvalue</span> <span class="op">&lt;</span> <span class="fl">.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## FALSE  TRUE </span></span>
<span><span class="co">##  1845   404</span></span></code></pre>
<p>A number of features may have the same <code>pvalue</code> and
<code>qvalue</code> due to the procedure for assessing significance. We
can additionally rank features by their effect size, to break ties in
the <code>qvalue</code>:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">most.sig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span>,</span>
<span>                 <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">qvalue</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">log2FC</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">most.sig</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"log2FC"</span>,<span class="st">"qvalue"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 2 columns</span></span>
<span><span class="co">##                      log2FC     qvalue</span></span>
<span><span class="co">##                   &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">## ENST00000264888.5  10.69374 7.2338e-05</span></span>
<span><span class="co">## ENST00000306602.2   9.34219 7.2338e-05</span></span>
<span><span class="co">## ENST00000306621.7   9.00874 7.2338e-05</span></span>
<span><span class="co">## ENST00000435136.6   8.11050 7.2338e-05</span></span>
<span><span class="co">## ENST00000307128.5  -6.96429 7.2338e-05</span></span>
<span><span class="co">## ENST00000382114.8  -5.92382 7.2338e-05</span></span></code></pre>
<p>The top 6 genes by <code>qvalue</code> and effect size in this case
all have positive LFC, although we have ranked by the largest in
absolute value, so large negative values will also appear in this
list.</p>
</div>
<div class="section level3">
<h3 id="plotting-results">Plotting results<a class="anchor" aria-label="anchor" href="#plotting-results"></a>
</h3>
<p>We can check the distribution of p-values. This looks as expected for
a comparison where we expect many transcripts will be affected by the
treatment (IFNg stimulation of macrophage cells). There is a flat
component and then an enrichment of transcripts with p-values near
0.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">pvalue</span>, col<span class="op">=</span><span class="st">"grey"</span><span class="op">)</span></span></code></pre></div>
<p><img src="swish_files/figure-html/unnamed-chunk-20-1.png"><!-- --></p>
<p>Of the transcripts in this set, which have the most extreme log2 fold
change? Note that often many transcripts will share the same q-value, so
it’s valuable to look at the log2 fold change as well (see further note
below on q-value computation). The log2 fold change computed by
<code>swish</code> is the median over inferential replicates, and uses a
pseudo-count of 5 on the scaled counts, to stabilize the variance on the
fold change from division by small counts. See the note above for
interpretation of log2 fold changes with respect to the levels of
<code>x</code>.</p>
<p>With the following code chunk, we construct two vectors that give the
significant genes with the lowest (most negative) and highest (most
positive) log2 fold changes.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span>,</span>
<span>     <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span>sig<span class="op">=</span><span class="va">qvalue</span> <span class="op">&lt;</span> <span class="fl">.05</span>, sign.lfc<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/sign.html" class="external-link">sign</a></span><span class="op">(</span><span class="va">log2FC</span><span class="op">)</span><span class="op">)</span></span>
<span>     <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        sign.lfc</span></span>
<span><span class="co">## sig      -1   1</span></span>
<span><span class="co">##   FALSE 998 847</span></span>
<span><span class="co">##   TRUE  184 220</span></span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sig</span> <span class="op">&lt;-</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">qvalue</span> <span class="op">&lt;</span> <span class="fl">.05</span></span>
<span><span class="va">lo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">log2FC</span> <span class="op">*</span> <span class="va">sig</span><span class="op">)</span></span>
<span><span class="va">hi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">log2FC</span> <span class="op">*</span> <span class="va">sig</span><span class="op">)</span></span></code></pre></div>
<p>Here we print a small table with just the calculated statistics for
the large positive log fold change transcripts (up-regulation):</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_up</span> <span class="op">&lt;-</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hi</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">top_up</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "tx_id"     "gene_id"   "tx_name"   "log10mean" "keep"      "stat"     </span></span>
<span><span class="co">##  [7] "log2FC"    "pvalue"    "locfdr"    "qvalue"</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"log10mean"</span>,<span class="st">"log2FC"</span>,<span class="st">"pvalue"</span>,<span class="st">"qvalue"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">top_up</span><span class="op">)</span><span class="op">[</span>,<span class="va">cols</span><span class="op">]</span>, digits<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                   log10mean log2FC   pvalue   qvalue</span></span>
<span><span class="co">## ENST00000264888.5      5.34  10.69 6.95e-06 7.23e-05</span></span>
<span><span class="co">## ENST00000306602.2      5.08   9.34 6.95e-06 7.23e-05</span></span>
<span><span class="co">## ENST00000306621.7      4.26   9.01 6.95e-06 7.23e-05</span></span>
<span><span class="co">## ENST00000435136.6      2.96   8.11 6.95e-06 7.23e-05</span></span>
<span><span class="co">## ENST00000502843.5      2.88   5.55 6.95e-06 7.23e-05</span></span>
<span><span class="co">## ENST00000500655.2      3.18   5.19 6.95e-06 7.23e-05</span></span></code></pre>
<p>Likewise for the largest negative log fold change transcripts
(down-regulation):</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_down</span> <span class="op">&lt;-</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">lo</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">top_down</span><span class="op">)</span><span class="op">[</span>,<span class="va">cols</span><span class="op">]</span>, digits<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                   log10mean log2FC   pvalue   qvalue</span></span>
<span><span class="co">## ENST00000307128.5      3.10  -6.96 6.95e-06 7.23e-05</span></span>
<span><span class="co">## ENST00000382114.8      2.97  -5.92 6.95e-06 7.23e-05</span></span>
<span><span class="co">## ENST00000381753.8      2.08  -3.49 4.04e-03 3.34e-02</span></span>
<span><span class="co">## ENST00000237612.7      2.79  -3.35 6.95e-06 7.23e-05</span></span>
<span><span class="co">## ENST00000407365.5      1.67  -3.00 5.70e-03 3.86e-02</span></span>
<span><span class="co">## ENST00000395002.6      2.67  -2.70 6.95e-06 7.23e-05</span></span></code></pre>
<p>We can plot the scaled counts for the inferential replicates, and
also group the samples by a covariate, in this case the cell line. The
analysis was paired, so the statistic assessed if the change within
pairs was consistent. Here we plot the 100th top up-regulated
transcript. <code>plotInfReps</code> also has functionality for plotting
uncertainty in single cell data, as will be covered in a later
section.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotInfReps.html">plotInfReps</a></span><span class="op">(</span><span class="va">y</span>, idx<span class="op">=</span><span class="va">hi</span><span class="op">[</span><span class="fl">100</span><span class="op">]</span>, x<span class="op">=</span><span class="st">"condition"</span>, cov<span class="op">=</span><span class="st">"line"</span><span class="op">)</span></span></code></pre></div>
<p><img src="swish_files/figure-html/unnamed-chunk-24-1.png"><!-- --></p>
<p>We can make an <a href="https://en.wikipedia.org/wiki/MA_plot" class="external-link">MA
plot</a>, where the transcripts in our FDR set are colored:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotMASwish.html">plotMASwish</a></span><span class="op">(</span><span class="va">y</span>, alpha<span class="op">=</span><span class="fl">.05</span><span class="op">)</span></span></code></pre></div>
<p><img src="swish_files/figure-html/unnamed-chunk-25-1.png"><!-- --></p>
<p>Using the <code>addIds</code> function from <em>tximeta</em>, we can
easily add gene symbols. By specifying <code>gene=TRUE</code>, this will
use the gene ID to match to gene symbols for all of the transcripts.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tximeta/man/addIds.html" class="external-link">addIds</a></span><span class="op">(</span><span class="va">y</span>, <span class="st">"SYMBOL"</span>, gene<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>We can then add gene symbols to our MA plot:</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotMASwish.html">plotMASwish</a></span><span class="op">(</span><span class="va">y</span>, alpha<span class="op">=</span><span class="fl">.05</span>, xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">.5</span>,<span class="fl">5.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span>, <span class="va">qvalue</span> <span class="op">&lt;</span> <span class="fl">.05</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">log2FC</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">4</span><span class="op">)</span>,</span>
<span>     <span class="fu"><a href="https://rdrr.io/r/graphics/text.html" class="external-link">text</a></span><span class="op">(</span><span class="va">log10mean</span>, <span class="va">log2FC</span>, <span class="va">SYMBOL</span>,</span>
<span>          col<span class="op">=</span><span class="st">"blue"</span>, pos<span class="op">=</span><span class="fl">4</span>, cex<span class="op">=</span><span class="fl">.7</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="swish_files/figure-html/unnamed-chunk-27-1.png"><!-- --></p>
</div>
</div>
<div class="section level2">
<h2 id="differential-gene-expression">Differential gene expression<a class="anchor" aria-label="anchor" href="#differential-gene-expression"></a>
</h2>
<div class="section level3">
<h3 id="running-swish-at-the-gene-level">Running Swish at the gene level<a class="anchor" aria-label="anchor" href="#running-swish-at-the-gene-level"></a>
</h3>
<p>We can also run swish at the gene level. First we summarize all of
the data to the gene level, using the <code>summarizeToGene</code>
function from <em>tximeta</em>. Again, we rename the object for
statistical analysis, and then we subset to the genes on chromosome 4
for the demonstration.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gse</span> <span class="op">&lt;-</span> <span class="fu">summarizeToGene</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="va">gy</span> <span class="op">&lt;-</span> <span class="va">gse</span></span>
<span><span class="va">gy</span> <span class="op">&lt;-</span> <span class="va">gy</span><span class="op">[</span><span class="fu">seqnames</span><span class="op">(</span><span class="va">gy</span><span class="op">)</span> <span class="op">==</span> <span class="st">"chr4"</span>,<span class="op">]</span></span></code></pre></div>
<p>Next we can run the same steps as before. Again we set a random seed
in order to be able to reproduce exact results in the future:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scaleInfReps.html">scaleInfReps</a></span><span class="op">(</span><span class="va">gy</span><span class="op">)</span></span>
<span><span class="va">gy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/labelKeep.html">labelKeep</a></span><span class="op">(</span><span class="va">gy</span><span class="op">)</span></span>
<span><span class="va">gy</span> <span class="op">&lt;-</span> <span class="va">gy</span><span class="op">[</span><span class="fu">mcols</span><span class="op">(</span><span class="va">gy</span><span class="op">)</span><span class="op">$</span><span class="va">keep</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">gy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/swish.html">swish</a></span><span class="op">(</span><span class="va">gy</span>, x<span class="op">=</span><span class="st">"condition"</span>, pair<span class="op">=</span><span class="st">"line"</span><span class="op">)</span></span></code></pre></div>
<p>As before, the number of genes in a 1% FDR set:</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">gy</span><span class="op">)</span><span class="op">$</span><span class="va">qvalue</span> <span class="op">&lt;</span> <span class="fl">.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## FALSE  TRUE </span></span>
<span><span class="co">##   403   226</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="plotting-gene-results">Plotting gene results<a class="anchor" aria-label="anchor" href="#plotting-gene-results"></a>
</h3>
<p>The histogram of p-values:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">pvalue</span>, col<span class="op">=</span><span class="st">"grey"</span><span class="op">)</span></span></code></pre></div>
<p><img src="swish_files/figure-html/unnamed-chunk-31-1.png"><!-- --></p>
<p>As before, finding the genes with the most extreme log2 fold
change:</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">gy</span><span class="op">)</span>,</span>
<span>     <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span>sig<span class="op">=</span><span class="va">qvalue</span> <span class="op">&lt;</span> <span class="fl">.05</span>, sign.lfc<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/sign.html" class="external-link">sign</a></span><span class="op">(</span><span class="va">log2FC</span><span class="op">)</span><span class="op">)</span></span>
<span>     <span class="op">)</span>     </span></code></pre></div>
<pre><code><span><span class="co">##        sign.lfc</span></span>
<span><span class="co">## sig      -1   1</span></span>
<span><span class="co">##   FALSE 251 152</span></span>
<span><span class="co">##   TRUE  110 116</span></span></code></pre>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sig</span> <span class="op">&lt;-</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">gy</span><span class="op">)</span><span class="op">$</span><span class="va">qvalue</span> <span class="op">&lt;</span> <span class="fl">.05</span></span>
<span><span class="va">glo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">gy</span><span class="op">)</span><span class="op">$</span><span class="va">log2FC</span> <span class="op">*</span> <span class="va">sig</span><span class="op">)</span></span>
<span><span class="va">ghi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="fu">mcols</span><span class="op">(</span><span class="va">gy</span><span class="op">)</span><span class="op">$</span><span class="va">log2FC</span> <span class="op">*</span> <span class="va">sig</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gtop_up</span> <span class="op">&lt;-</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">gy</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ghi</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">gtop_up</span><span class="op">)</span><span class="op">[</span>,<span class="va">cols</span><span class="op">]</span>, digits<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                    log10mean log2FC   pvalue   qvalue</span></span>
<span><span class="co">## ENSG00000138755.5       5.36  10.73 2.48e-05 9.89e-05</span></span>
<span><span class="co">## ENSG00000169245.5       5.10   9.37 2.48e-05 9.89e-05</span></span>
<span><span class="co">## ENSG00000169248.12      4.28   8.95 2.48e-05 9.89e-05</span></span>
<span><span class="co">## ENSG00000004468.12      3.61   5.55 2.48e-05 9.89e-05</span></span>
<span><span class="co">## ENSG00000145365.10      3.87   5.16 2.48e-05 9.89e-05</span></span>
<span><span class="co">## ENSG00000171509.15      2.42   4.68 2.48e-05 9.89e-05</span></span></code></pre>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gtop_down</span> <span class="op">&lt;-</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">gy</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">glo</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">gtop_down</span><span class="op">)</span><span class="op">[</span>,<span class="va">cols</span><span class="op">]</span>, digits<span class="op">=</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                    log10mean log2FC   pvalue   qvalue</span></span>
<span><span class="co">## ENSG00000172399.5       3.12  -6.96 2.48e-05 9.89e-05</span></span>
<span><span class="co">## ENSG00000153012.11      3.00  -5.87 2.48e-05 9.89e-05</span></span>
<span><span class="co">## ENSG00000118777.11      2.80  -3.19 2.48e-05 9.89e-05</span></span>
<span><span class="co">## ENSG00000151725.11      2.83  -2.48 2.48e-05 9.89e-05</span></span>
<span><span class="co">## ENSG00000248187.1       2.27  -2.41 2.48e-05 9.89e-05</span></span>
<span><span class="co">## ENSG00000145390.11      3.59  -2.36 2.48e-05 9.89e-05</span></span></code></pre>
<p>We can plot a particular one of these genes:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotInfReps.html">plotInfReps</a></span><span class="op">(</span><span class="va">gy</span>, idx<span class="op">=</span><span class="va">ghi</span><span class="op">[</span><span class="fl">100</span><span class="op">]</span>, x<span class="op">=</span><span class="st">"condition"</span>, cov<span class="op">=</span><span class="st">"line"</span><span class="op">)</span></span></code></pre></div>
<p><img src="swish_files/figure-html/unnamed-chunk-34-1.png"><!-- --></p>
<p>As expected, the highly up-regulated genes are involved in immune
response. Many genes encoding guanylate-binding proteins (GBP) are
up-regulated, and these proteins are induced by interferon, produced in
response to infection by pathogenic microbes.</p>
<p>We can make an MA plot, where the genes in our FDR set are
colored:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotMASwish.html">plotMASwish</a></span><span class="op">(</span><span class="va">gy</span>, alpha<span class="op">=</span><span class="fl">.05</span><span class="op">)</span></span></code></pre></div>
<p><img src="swish_files/figure-html/unnamed-chunk-35-1.png"><!-- --></p>
<p>Again, using the <code>addIds</code> function from <em>tximeta</em>,
we can easily add gene symbols to our gene-level expression
analysis:</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="va">gy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tximeta/man/addIds.html" class="external-link">addIds</a></span><span class="op">(</span><span class="va">gy</span>, <span class="st">"SYMBOL"</span>, gene<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>We can then add gene symbols to our MA plot:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotMASwish.html">plotMASwish</a></span><span class="op">(</span><span class="va">gy</span>, alpha<span class="op">=</span><span class="fl">.05</span>, xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">.5</span>,<span class="fl">5.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">gy</span><span class="op">)</span>, <span class="va">qvalue</span> <span class="op">&lt;</span> <span class="fl">.05</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">log2FC</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">3</span><span class="op">)</span>,</span>
<span>     <span class="fu"><a href="https://rdrr.io/r/graphics/text.html" class="external-link">text</a></span><span class="op">(</span><span class="va">log10mean</span>, <span class="va">log2FC</span>, <span class="va">SYMBOL</span>,</span>
<span>          col<span class="op">=</span><span class="st">"blue"</span>, pos<span class="op">=</span><span class="fl">4</span>, cex<span class="op">=</span><span class="fl">.7</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="swish_files/figure-html/unnamed-chunk-37-1.png"><!-- --></p>
</div>
</div>
<div class="section level2">
<h2 id="differential-transcript-usage">Differential transcript usage<a class="anchor" aria-label="anchor" href="#differential-transcript-usage"></a>
</h2>
<p>We have added a new function <code>isoformProportions</code> which
can be run after <code>scaleInfReps</code> (and optionally after
removing genes via <code>labelKeep</code> and subsetting the
SummarizedExperiment). This function, <code>isoformProportions</code>
will create a new assay <code>isoProp</code> from the scaledTPM counts,
containing isoform proportions per gene. The same procedure will also be
applied to all the inferential replicates. Note that after
<code>isoformProportions</code> the transcripts from single isoform
genes will be removed, and the transcripts will be re-ordered by gene
(alphabetically by gene).</p>
<p>Following this function, running <code>swish</code> will be
equivalent to a test of differential transcript usage, taking account of
the uncertainty in transcript abundances, as it will look for
transcripts where the isoform proportions change across condition.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># run on the transcript-level dataset</span></span>
<span><span class="va">iso</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/isoformProportions.html">isoformProportions</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="va">iso</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/swish.html">swish</a></span><span class="op">(</span><span class="va">iso</span>, x<span class="op">=</span><span class="st">"condition"</span>, pair<span class="op">=</span><span class="st">"line"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="interaction-designs">Interaction designs<a class="anchor" aria-label="anchor" href="#interaction-designs"></a>
</h2>
<p>We also provide in <code>swish</code> methods for testing if a
condition effect varies <em>across a secondary covariate</em>, using
matched samples for condition, or un-matched samples, which we refer to
as “interactions” in the software.</p>
<p>If matched samples are available, we compute the log2 fold change for
each pair of samples across condition in the same covariate group, and
then we use a Wilcoxon rank sum statistic for comparing the log2 fold
changes across the secondary covariate. For permutation significance,
the secondary covariate labels of the pairs are permuted. For un-matched
samples, multiple random “pseudo-pairs” of samples across condition
within the two covariate groups are chosen, and the statistic computed
as above, averaging over the random pseudo-pairings. The motivation for
the above permutation schemes is to ensure the following condition, that
“under the null hypothesis, the likelihood of the data is invariant
under these permutations” <span class="citation">(Anderson and Ter Braak
2003)</span>, where our null hypothesis specifically involves the
interaction between condition and the secondary covariate.</p>
<p>For the macrophage dataset we have been working with <span class="citation">(Alasoo et al. 2018)</span>, we have a 2x2 experimental
design, with IFN gamma stimulation, <em>Salmonella</em> infection, and
both treatments, as well as control samples. We have these four
conditions across 6 cell lines from 6 donors (a subset of all the
RNA-seq samples available). So we can use the first method described
above, where the cell line is used to match samples across
condition.</p>
<div class="section level3">
<h3 id="condition-and-secondary-covariates">Condition and secondary covariates<a class="anchor" aria-label="anchor" href="#condition-and-secondary-covariates"></a>
</h3>
<p>We begin the interaction analysis by now loading in the full set of
samples, as before we only loaded the samples for the two group
analysis. We then define two new factors indicating IFNg status and
<em>Salmonella</em> status.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coldata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"coldata.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">coldata</span> <span class="op">&lt;-</span> <span class="va">coldata</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">5</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"names"</span>,<span class="st">"id"</span>,<span class="st">"line"</span>,<span class="st">"condition"</span><span class="op">)</span></span>
<span><span class="va">coldata</span><span class="op">$</span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>, <span class="st">"quants"</span>, <span class="va">coldata</span><span class="op">$</span><span class="va">names</span>, <span class="st">"quant.sf.gz"</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tximeta/man/tximeta.html" class="external-link">tximeta</a></span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span><span class="op">$</span><span class="va">ifng</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"IFNg"</span>,<span class="va">se</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span>,</span>
<span>  <span class="st">"treated"</span>,<span class="st">"control"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">se</span><span class="op">$</span><span class="va">salmonella</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"SL1344"</span>,<span class="va">se</span><span class="op">$</span><span class="va">condition</span><span class="op">)</span>,</span>
<span>  <span class="st">"infected"</span>,<span class="st">"control"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span>,</span>
<span>     <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">ifng</span>, <span class="va">salmonella</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          salmonella</span></span>
<span><span class="co">## ifng      control infected</span></span>
<span><span class="co">##   control       6        6</span></span>
<span><span class="co">##   treated       6        6</span></span></code></pre>
<p>We will work with the chromosome 1 transcripts for demonstration:</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y2</span> <span class="op">&lt;-</span> <span class="va">se</span></span>
<span><span class="va">y2</span> <span class="op">&lt;-</span> <span class="va">y2</span><span class="op">[</span><span class="fu">seqnames</span><span class="op">(</span><span class="va">y2</span><span class="op">)</span> <span class="op">==</span> <span class="st">"chr1"</span>,<span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-and-check-paired-samples">Create and check paired samples<a class="anchor" aria-label="anchor" href="#create-and-check-paired-samples"></a>
</h3>
<p>Our implementation of the interaction design for matched samples
takes into account matched samples within the <code>x</code> condition,
which we will specify to be the <em>Salmonella</em> infection status. We
will specify the secondary covariate <code>cov</code> to be the IFN
gamma treatment. We will look for transcripts where the infection
response changes based on IFN gamma treatment.</p>
<p>To perform the analysis, we create a new variable <code>pair</code>
which will record which samples are related within groups defined by IFN
gamma treatment status.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y2</span><span class="op">$</span><span class="va">pair</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">y2</span><span class="op">$</span><span class="va">line</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">y2</span><span class="op">$</span><span class="va">pair</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">LETTERS</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span> <span class="co"># simplify names</span></span>
<span><span class="va">y2</span><span class="op">$</span><span class="va">pair</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">y2</span><span class="op">$</span><span class="va">pair</span><span class="op">)</span></span>
<span><span class="va">y2</span><span class="op">$</span><span class="va">pair</span><span class="op">[</span><span class="va">y2</span><span class="op">$</span><span class="va">ifng</span> <span class="op">==</span> <span class="st">"control"</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "A" "A" "B" "B" "C" "C" "D" "D" "E" "E" "F" "F"</span></span></code></pre>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y2</span><span class="op">$</span><span class="va">pair</span><span class="op">[</span><span class="va">y2</span><span class="op">$</span><span class="va">ifng</span> <span class="op">==</span> <span class="st">"treated"</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "A" "A" "B" "B" "C" "C" "D" "D" "E" "E" "F" "F"</span></span></code></pre>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y2</span><span class="op">$</span><span class="va">pair</span><span class="op">[</span><span class="va">y2</span><span class="op">$</span><span class="va">ifng</span> <span class="op">==</span> <span class="st">"treated"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">LETTERS</span><span class="op">[</span><span class="fl">7</span><span class="op">:</span><span class="fl">12</span><span class="op">]</span>, each<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">y2</span><span class="op">$</span><span class="va">pair</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">y2</span><span class="op">$</span><span class="va">pair</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">y2</span><span class="op">$</span><span class="va">pair</span>, <span class="va">y2</span><span class="op">$</span><span class="va">salmonella</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    </span></span>
<span><span class="co">##     control infected</span></span>
<span><span class="co">##   A       1        1</span></span>
<span><span class="co">##   B       1        1</span></span>
<span><span class="co">##   C       1        1</span></span>
<span><span class="co">##   D       1        1</span></span>
<span><span class="co">##   E       1        1</span></span>
<span><span class="co">##   F       1        1</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="swish-for-interaction-effects">Swish for interaction effects<a class="anchor" aria-label="anchor" href="#swish-for-interaction-effects"></a>
</h3>
<p>We now perform <code>swish</code> analysis, specifying the
<em>Salmonella</em> infection as our main condition, the IFN gamma
treatment as the secondary covariate, and providing the pairing within
IFN gamma treatment groups. We specify <code>interaction=TRUE</code> to
test for differences in infection response across IFN gamma treatment
group.</p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scaleInfReps.html">scaleInfReps</a></span><span class="op">(</span><span class="va">y2</span><span class="op">)</span></span>
<span><span class="va">y2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/labelKeep.html">labelKeep</a></span><span class="op">(</span><span class="va">y2</span><span class="op">)</span></span>
<span><span class="va">y2</span> <span class="op">&lt;-</span> <span class="va">y2</span><span class="op">[</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y2</span><span class="op">)</span><span class="op">$</span><span class="va">keep</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">y2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/swish.html">swish</a></span><span class="op">(</span><span class="va">y2</span>, x<span class="op">=</span><span class="st">"salmonella"</span>, cov<span class="op">=</span><span class="st">"ifng"</span>, pair<span class="op">=</span><span class="st">"pair"</span>, interaction<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="plotting-interaction-results">Plotting interaction results<a class="anchor" aria-label="anchor" href="#plotting-interaction-results"></a>
</h3>
<p>In this case, we appear to have fewer non-null p-values from first
impression of the p-value histogram:</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y2</span><span class="op">)</span><span class="op">$</span><span class="va">pvalue</span>, col<span class="op">=</span><span class="st">"grey"</span><span class="op">)</span></span></code></pre></div>
<p><img src="swish_files/figure-html/unnamed-chunk-44-1.png"><!-- --></p>
<p>The MA plot shows significant transcripts on either side of
<code>log2FC=0</code>. Note that the log2 fold change reported is the
<em>difference</em> between the log2 fold change in the IFN gamma
treated and IFN gamma control group. So positive <code>log2FC</code> in
this plot indicates that the effect is higher with IGN gamma treatment
than in absence of the treatment.</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotMASwish.html">plotMASwish</a></span><span class="op">(</span><span class="va">y2</span>, alpha<span class="op">=</span><span class="fl">.05</span><span class="op">)</span></span></code></pre></div>
<p><img src="swish_files/figure-html/unnamed-chunk-45-1.png"><!-- --></p>
<p>We can plot some of the transcripts with high log fold change
<em>difference</em> across IFN gamma treatment group, and which belong
to the less than 5% nominal FDR group:</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">qvalue</span> <span class="op">&lt;</span> <span class="fl">.05</span> <span class="op">&amp;</span> <span class="va">log2FC</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotInfReps.html">plotInfReps</a></span><span class="op">(</span><span class="va">y2</span>, <span class="va">idx</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, x<span class="op">=</span><span class="st">"ifng"</span>, cov<span class="op">=</span><span class="st">"salmonella"</span><span class="op">)</span></span></code></pre></div>
<p><img src="swish_files/figure-html/unnamed-chunk-46-1.png"><!-- --></p>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotInfReps.html">plotInfReps</a></span><span class="op">(</span><span class="va">y2</span>, <span class="va">idx</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, x<span class="op">=</span><span class="st">"ifng"</span>, cov<span class="op">=</span><span class="st">"salmonella"</span><span class="op">)</span></span></code></pre></div>
<p><img src="swish_files/figure-html/unnamed-chunk-46-2.png"><!-- --></p>
</div>
</div>
<div class="section level2">
<h2 id="allelic-expression-analysis">Allelic expression analysis<a class="anchor" aria-label="anchor" href="#allelic-expression-analysis"></a>
</h2>
<p>We have a separate vignette for allelic expression analysis, either
navigate to the <a href="https://mikelove.github.io/fishpond/articles/allelic.html">allelic
vignette</a> on the <em>fishpond</em> website, or the following from the
R prompt:</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/vignette.html" class="external-link">vignette</a></span><span class="op">(</span>topic<span class="op">=</span><span class="st">"allelic"</span>, package<span class="op">=</span><span class="st">"fishpond"</span><span class="op">)</span></span></code></pre></div>
<p>The allelic analysis methods are described in <span class="citation">Wu et al. (2022)</span> <a href="https://doi.org/10.1101/2022.08.12.503785" class="external-link">doi:
10.1101/2022.08.12.503785</a>.</p>
</div>
<div class="section level2">
<h2 id="correlation-test">Correlation test<a class="anchor" aria-label="anchor" href="#correlation-test"></a>
</h2>
<p><em>Swish</em> now has methods to compute correlations
(<code>"spearman"</code> or <code>"pearson"</code>) of a continuous
variable <code>x</code> with the log counts, and then assess the
significance of those correlations using inferential replicates and
permutation tests. Additionally, one can compute correlations of a
covariate <code>cov</code> with log fold changes across a pairing
variable (where <code>x</code> labels the condition for the log fold
change and <code>pair</code> labels the pairs). The Spearman-based test
statistic follows the procedure from the original method <em>SAMseq</em>
that <em>Swish</em> is based upon <span class="citation">(Li and
Tibshirani 2011)</span>, e.g. the methods used in that package with a
quantitative response type. See <code><a href="../reference/swish.html">?swish</a></code> and the
<code>cor</code> argument for more details on the use of correlations
within <em>Swish</em>.</p>
<p>For examples of the correlation tests, see the dynamic AI examples in
the <code>"allelic"</code> fishpond vignette, or there are also some
simulated examples in the <code>test_correlation.R</code> testing script
located in fishpond’s <code>tests/testthat</code> directory.</p>
</div>
<div class="section level2">
<h2 id="alevin-scrna-seq">alevin scRNA-seq<a class="anchor" aria-label="anchor" href="#alevin-scrna-seq"></a>
</h2>
<p>The <em>alevin</em> <span class="citation">(Srivastava et al.
2019)</span> and <em>tximport</em> / <em>tximeta</em> maintainers have
created an efficient format for storing and importing the sparse
scRNA-seq estimated gene counts, inferential mean and variance data, and
optionally the inferential replicate counts. <code>tximeta</code> will
automatically import these matrices if <em>alevin</em> was run using
<code>--numCellBootstraps</code> (in order to generate inferential
variance) and additionally <code>--dumpFeatures</code> (in order to dump
the inferential replicates themselves, see below on thoughts on avoiding
this step though).</p>
<p>The storage format for counts, mean, variance, and inferential
replicates, involves writing one cell at a time, storing the locations
of the non-zero counts, and then the non-zero counts. The matrices are
loaded into R sparely using the <em>Matrix</em> package. The storage
format is efficient, for example, the estimated counts for the 900 mouse
neuron dataset from 10x Genomics takes up 4.2 Mb, the mean/variance
matrices take up 8.6 Mb each, and the inferential replicates takes up 72
Mb (20 bootstrap inferential replicates). Hence avoiding writing and
importing the inferential replicates themselves, and only using the mean
and variance, is desirable.</p>
<p>The <code>swish</code> and <em>alevin</em> authors have developed a
workflow that uses inferential mean and variance to generate
pseudo-inferential replicates in place of the actual inferential
replicates. Storing and importing only the inferential mean and variance
is much more efficient (dramatically faster load time and less space on
disk and in memory). The faster workflow would then skip
<code>--dumpFeatures</code> when running <em>alevin</em>, or
subsequently use <code>dropInfReps=TRUE</code> to not load the
inferential replicates into R.</p>
<p><strong>Plotting:</strong> To demonstrate how the inferential mean
and variance look across real scRNA-seq data, we load the neurons
dataset and plot the inferential replicate data across cells. First we
read in the counts, in this case using <code>dropInfReps=TRUE</code>
because the directory includes the actual inferential replicates, not
only the mean and variance. We set <code>skipMeta=TRUE</code>, although
in general you would want <code>tximeta</code> to add the gene range
information and other metadata if working with human, mouse, or fruit
fly.</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"tximportData"</span><span class="op">)</span></span>
<span><span class="va">files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">dir</span>,<span class="st">"alevin/neurons_900_v014/alevin/quants_mat.gz"</span><span class="op">)</span></span>
<span><span class="va">neurons</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/tximeta/man/tximeta.html" class="external-link">tximeta</a></span><span class="op">(</span><span class="va">files</span>, type<span class="op">=</span><span class="st">"alevin"</span>,</span>
<span>                   skipMeta<span class="op">=</span><span class="cn">TRUE</span>, <span class="co"># just for vignette</span></span>
<span>                   dropInfReps<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                   alevinArgs<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>filterBarcodes<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We can easily make a <em>SingleCellExperiment</em> object <span class="citation">(Amezquita et al. 2020)</span>, and plot counts across
cluster (here a randomly assigned cluster label). For more details on
working with <em>SingleCellExperiment</em> objects, consult the
following online book: <a href="https://osca.bioconductor.org/" class="external-link">Orchestrating Single-Cell Analysis
with Bioconductor</a> <span class="citation">(Amezquita et al.
2020)</span>.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">neurons</span>, <span class="st">"SingleCellExperiment"</span><span class="op">)</span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cl"</span>,<span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>,<span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>,<span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span>, mar<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">2</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotInfReps.html">plotInfReps</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"ENSMUSG00000072235.6"</span>, x<span class="op">=</span><span class="st">"cluster"</span>,</span>
<span>            legend<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plotInfReps.html">plotInfReps</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"ENSMUSG00000072235.6"</span>, x<span class="op">=</span><span class="st">"cluster"</span>,</span>
<span>            reorder<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><code>plotInfReps</code> has a number of options and convenience
arguments for visualizing single cell data. One can:</p>
<ul>
<li>add a <code>legend</code>,</li>
<li>
<code>reorder</code> the cells by expression value within
cluster,</li>
<li>apply size factors to the counts (<code>applySF</code>) (size factor
scaling will use the values in <code>sizeFactors(sce)</code> or
equivalently <code>mcols(sce)$sizeFactor</code>),</li>
<li>use a column of <code>mcols(sce)</code> to set the <code>main</code>
title, e.g. <code>mainCol="SYMBOL"</code>,</li>
<li>specify <code>x</code> as a numeric covariate (e.g. pseudotime), and
use <code>cov</code> to distinguish groups (e.g. lineage). Points will
then be colored by <code>cov</code> instead of by discrete
<code>x</code>.</li>
</ul>
<p>See <code><a href="../reference/plotInfReps.html">?plotInfReps</a></code> for more description of arguments.</p>
<p>Note that the figures scale to some degree by the number of cells;
for example with <span class="math inline">\(n \in [1,150)\)</span> or
<span class="math inline">\([150,400)\)</span>, more visual elements per
cell will be included:</p>
<p><strong>Advice on Swish testing:</strong> <code>swish</code> can be
run on <em>alevin</em> counts imported with <code>tximeta</code>, but
there are a few extra steps suggested. The following does not use
evaluated code chunks, but provides suggestions for how to tailor
<code>swish</code> to single-cell datasets.</p>
<p><strong>Filter genes:</strong> we recommend to filter genes as the
first step, to reduce the size of the data before losing sparsity on the
count matrices (conversion of data to ranks loses data sparsity inside
the <code><a href="../reference/swish.html">swish()</a></code> function). One can run <code>labelKeep</code>
therefore before <code>scaleInfReps</code>. E.g., to remove genes for
which there are not 10 cells with a count of 3 or more:</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/labelKeep.html">labelKeep</a></span><span class="op">(</span><span class="va">y</span>, minCount<span class="op">=</span><span class="fl">3</span>, minN<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">keep</span>,<span class="op">]</span> <span class="co"># subset genes</span></span></code></pre></div>
<p><strong>Subset cells:</strong> One should also subset to the groups
of cells of interest for differential testing, in order to take up the
least amount of memory when the sparse matrices are converted to dense
matrices. Note that <code>swish</code> only allows for differential
testing of two groups (although it allows for blocking factors and
interaction tests).</p>
<p><strong>(Slower) With inferential replicates:</strong> After one has
filtered both genes and cells down to the set that are of interest for
differential testing, one can run the following commands, to make the
sparse matrices into dense ones, scale the cells, and perform
<em>Swish</em> differential expression, however read on for faster
suggestions.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span>, <span class="va">as.matrix</span><span class="op">)</span> <span class="co"># make dense matrices</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scaleInfReps.html">scaleInfReps</a></span><span class="op">(</span><span class="va">y</span>, lengthCorrect<span class="op">=</span><span class="cn">FALSE</span>, sfFun<span class="op">=</span><span class="va">sfFun</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/swish.html">swish</a></span><span class="op">(</span><span class="va">y</span>, x<span class="op">=</span><span class="st">"condition"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Size factor function:</strong> Note that
<code>scaleInfReps</code> has an argument <code>sfFun</code> which
allows the user to provide their own size factor calculation function.
We have found that <code>calculateSumFactors</code> <span class="citation">(Aaron T. L. Lun, Bach, and Marioni 2016)</span> in the
<em>scran</em> package <span class="citation">(A. T. L. Lun, McCarthy,
and Marioni 2016)</span> works well for computing size factors.
<code>sfFun</code> should be specified as a function that returns a
vector of size factors, or a numeric vector of the size factors
themselves.</p>
<p><strong>(Faster) Without inferential replicates:</strong> The
following workflow can be used in the case that
<code>assayNames(y)</code> only contains counts, mean, and variance,
which is much faster by avoiding writing/importing inferential
replicates. We first generate pseudo-inferential replicates from
inferential mean and variance matrices before running
<code>scaleInfReps</code> from the code chunk above. The generation of
pseudo-inferential replicates is described in <span class="citation">Van
Buren et al. (2020)</span>.</p>
<p>The following function can be used just before
<code>scaleInfReps</code>:</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeInfReps.html">makeInfReps</a></span><span class="op">(</span><span class="va">y</span>, <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<p>There is alternatively a scheme for splitting the operation of
generating (dense) inferential replicate matrices across multiple jobs,
and running <code>swish</code> across batches of genes at a time. This
job-splitting procedure is also described and benchmarked in <span class="citation">Van Buren et al. (2020)</span>. This helps to reduce
the total memory used, in the case that the counts, mean, and variance
matrices are too large to be made dense altogether. This scheme involves
1) splitting the object into smaller pieces, written out to
<code>.rds</code> files, 2) running <code>swish</code> as a distributed
job, 3) reading the <code>.csv</code> output back into R. The following
code chunk would start with a ‘y’ with sparse matrix assays, and without
ever running <code>scaleInfReps</code> (it is run within the distributed
job, and with <code>lengthCorrect=FALSE</code> by default).</p>
<div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">y</span>, <span class="st">"SingleCellExperiment"</span><span class="op">)</span></span>
<span><span class="co"># then, after filtering genes and cells...</span></span>
<span><span class="co"># compute and store sizeFactors calculated over all genes</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">scran</span><span class="fu">::</span><span class="fu">computeSumFactors</span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="co"># split swish objects into 8 RDS files:</span></span>
<span><span class="fu"><a href="../reference/splitSwish.html">splitSwish</a></span><span class="op">(</span><span class="va">y</span>, nsplits<span class="op">=</span><span class="fl">8</span>, prefix<span class="op">=</span><span class="st">"swish"</span>, snakefile<span class="op">=</span><span class="st">"Snakefile"</span><span class="op">)</span></span>
<span><span class="co"># now, run snakemake from command line</span></span>
<span><span class="co"># after finished, results back into R:</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addStatsFromCSV.html">addStatsFromCSV</a></span><span class="op">(</span><span class="va">y</span>, <span class="st">"summary.csv"</span><span class="op">)</span></span></code></pre></div>
<p>The <code>splitSwish</code> function will write out a
<code>Snakefile</code> that can be used with <a href="https://snakemake.readthedocs.io/en/stable/" class="external-link">snakemake</a> in
order to run distributed <code>swish</code> jobs in an easily customized
workflow. Then the <code>addStatsFromCSV</code> will read in and attach
the results to the original object. This final alternative avoids
generating dense matrices until they have been split into
<code>nsplits</code> pieces, and so can be used to reduce the memory
requirements for the individual jobs. If one is new to running
<code>snakemake</code>, it is recommended to first run with the flags
<code>-np</code> as a “dry-run” to see the operations that will be
performed. The <code>swish</code> command can be customized in the
<code>swish</code> rule in the <code>Snakefile</code>, e.g. to control
for batches or test for interactions.</p>
</div>
<div class="section level2">
<h2 id="further-details">Further details<a class="anchor" aria-label="anchor" href="#further-details"></a>
</h2>
<div class="section level3">
<h3 id="analysis-types-supported-by-swish">Analysis types supported by Swish<a class="anchor" aria-label="anchor" href="#analysis-types-supported-by-swish"></a>
</h3>
<p>There are currently seven types of analysis supported by
<code>swish</code>:</p>
<ul>
<li>Two group analysis</li>
<li>Two groups with two or more batches</li>
<li>Two group paired or matched samples</li>
<li>Two condition x two group paired samples, interaction test</li>
<li>Two condition x two group samples, not paired, interaction test</li>
<li>Correlation test of expression with continuous x</li>
<li>Correlation test of LFC for paired samples with continuous
covariate</li>
</ul>
<p>This vignette demonstrated the third in this list, but the others can
be run by either not specifying any additional covariates, or by
specifying a batch variable with the argument <code>cov</code> instead
of <code>pair</code>. The two interaction tests can be run by specifying
<code>interaction=TRUE</code> and providing <code>x</code>,
<code>cov</code>, and optionally <code>pair</code>. The last two tests
can be run using the <code>cor</code> argument (see <a href="#correlation-test">Correlation test</a> section in this
vignette).</p>
</div>
<div class="section level3">
<h3 id="accounting-for-continuous-variables">Accounting for continuous variables<a class="anchor" aria-label="anchor" href="#accounting-for-continuous-variables"></a>
</h3>
<p>We have two recommended approaches for using <code>swish</code> in
combination with estimated batch effects, e.g. factors of unwanted
variation estimated by <em>RUVSeq</em> or surrogate variables estimated
by <em>SVA</em> (without loss of generality, we will call these batch
factors). First, examine a plot of the estimated batch factors, e.g. a
stripplot (1 factor) or scatterplot of pairs (2 or more factors).</p>
<ol style="list-style-type: decimal">
<li>If the samples seem to fall into discrete clusters, one can run
<code>kmeans</code> to assign a discrete estimated batch to each
cluster. If the number of samples per condition group per batch is 2 or
more, then one can use the <code>cluster</code> output of the
<code>kmeans</code> function as input to the <code>cov</code> argument
of <code>swish</code>. This will then perform a stratified Mann-Whitney
Wilcoxon test.</li>
<li>If the samples do not fall into discrete clusters, one can use the
same approach that we use to correct the samples for sequencing depth
variation: direct scaling of the estimated counts across the inferential
replicates. Why this works: usually we do not scale counts in
statistical inference pipelines because we lose the information about
precision of counts across the dynamic range. However, because we keep
the set of inferential replicate matrices, and these contain both
sampling and additional inferential variance, we are able to track how
scaling the counts affects the variance, and this informs the test
statistic.</li>
</ol>
<p>For the second approach, one can directly scale the inferential
counts using <em>limma</em> and <code>removeBatchEffect</code>.</p>
<p>Here a demonstration using simulated data:</p>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeSimSwishData.html">makeSimSwishData</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>First we perform standard scaling for sequencing depth and labeling
of feautures with too few counts. We save the mean of scaled inferential
replicates using <code>saveMeanScaled=TRUE</code> as we will use this to
estimate the batch factors.</p>
<div class="sourceCode" id="cb79"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/scaleInfReps.html">scaleInfReps</a></span><span class="op">(</span><span class="va">y</span>, saveMeanScaled<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span>,<span class="fl">4</span><span class="op">)</span> <span class="co"># 'meanScaled' assay</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "infRep18"   "infRep19"   "infRep20"   "meanScaled"</span></span></code></pre>
<div class="sourceCode" id="cb81"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/labelKeep.html">labelKeep</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">keep</span>,<span class="op">]</span></span></code></pre></div>
<p>Next estimate the batch factors on the mean of scaled inferential
count matrices.</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">norm_cts_for_batch</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">y</span>, <span class="st">"meanScaled"</span><span class="op">)</span></span>
<span><span class="co"># use batch factor estimation method of your choosing</span></span>
<span><span class="co"># ...</span></span></code></pre></div>
<p>Suppose we estimate two batch factors, <code>w1</code> and
<code>w2</code>, now we remove the variation from each inferential
replicate associated with these, using limma’s
<code>removeBatchEffect</code>.</p>
<div class="sourceCode" id="cb83"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">w1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="co"># here simulated w1, use real instead</span></span>
<span><span class="va">w2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span> <span class="co"># here simulated w2, use real instead</span></span>
<span><span class="va">W</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">w1</span>, <span class="va">w2</span><span class="op">)</span></span>
<span><span class="va">infRepIdx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"infRep"</span>,<span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span>,value<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">nreps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">infRepIdx</span><span class="op">)</span></span></code></pre></div>
<p>The following loads <em>limma</em> and removes batch associated
variation on the log2 scale (a pseudocount is added to avoid
<code>-Inf</code> values).</p>
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/limma/" class="external-link">limma</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="va">condition</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pc</span> <span class="op">&lt;-</span> <span class="fl">.1</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">k</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">nreps</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">logInfRep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">infRepIdx</span><span class="op">[</span><span class="va">k</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span> <span class="va">pc</span><span class="op">)</span></span>
<span>  <span class="va">logInfRep</span> <span class="op">&lt;-</span> <span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/removeBatchEffect.html" class="external-link">removeBatchEffect</a></span><span class="op">(</span></span>
<span>                        <span class="va">logInfRep</span>,</span>
<span>                        covariates<span class="op">=</span><span class="va">W</span>,</span>
<span>                        design<span class="op">=</span><span class="va">mm</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">infRepIdx</span><span class="op">[</span><span class="va">k</span><span class="op">]</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="va">logInfRep</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Now we run <code>swish</code> as usual:</p>
<div class="sourceCode" id="cb85"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/swish.html">swish</a></span><span class="op">(</span><span class="va">y</span>, x<span class="op">=</span><span class="st">"condition"</span><span class="op">)</span></span></code></pre></div>
<p>While we have assessed this approach with a small number of estimated
nuisance covariates, note that with many covariates this approach may
lead to inflation of test statistics.</p>
</div>
<div class="section level3">
<h3 id="structure-of-tximeta-output">Structure of tximeta output<a class="anchor" aria-label="anchor" href="#structure-of-tximeta-output"></a>
</h3>
<p>While <code>tximeta</code> is the safest way to provide the correct
input to <code>swish</code>, all that <code>swish</code> requires for
running is a <em>SummarizedExperiment</em> object with the following
assays: <code>counts</code>, <code>length</code>, and
<code>infRep1</code>, <code>infRep2</code>, …, <code>infRepN</code>,
where <code>N</code> is simply the number of Gibbs samples or boostraps
samples, e.g. 20 in the examples above. The counts and inferential
replicates are estimated counts from a quantification method, either at
the transcript level or summed to the gene level (simple sum). These
counts sum up to the (mapped) library size for each sample. It is
assumed that the <code>length</code> matrix gives the effective lengths
for each transcript, or average transcript length for each gene as
summarized by the functions in
<code>tximeta</code>/<code>tximport</code>. If the counts should not be
corrected for effective length (e.g. 3’ tagged RNA-seq), then
<code>lengthCorrect=FALSE</code> should be specified when running
<code>scaleInfReps</code>.</p>
<p>Note on simulation: it is difficult to simulate inferential
uncertainty in a realistic manner without construction of reads from
transcripts, using a method like <em>polyester</em>. Constructing reads
from the reference transcriptome or a sample-specific transcriptome
naturally produces the structure of read-assignment inferential
uncertainty that <code>swish</code> and other methods control for in
real RNA-seq data.</p>
</div>
<div class="section level3">
<h3 id="plotting-q-values-over-statistics">Plotting q-values over statistics<a class="anchor" aria-label="anchor" href="#plotting-q-values-over-statistics"></a>
</h3>
<p>As with <em>SAMseq</em> and <em>SAM</em>, <code>swish</code> makes
use of the permutation plug-in approach for q-value calculation.
<code>swish</code> calls the <code>empPvals</code> and
<code>qvalue</code> functions from the <em>qvalue</em> package to
calculate the q-values (or optionally similar functions from the
<em>samr</em> package). If we plot the q-values against the statistic,
or against the log2 fold change, one can see clusters of genes with the
same q-value (because they have the same or similar statistic). One
consequence of this is that, in order to rank the genes, rather than
ranking directly by q-value, it makes more sense to pick a q-value
threshold and then within that set of genes, to rank by the log fold
change, as shown above when the code chunk has
<code>log2FC * sig</code>.</p>
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gres</span> <span class="op">&lt;-</span> <span class="fu">mcols</span><span class="op">(</span><span class="va">gy</span><span class="op">)</span><span class="op">[</span><span class="fu">mcols</span><span class="op">(</span><span class="va">gy</span><span class="op">)</span><span class="op">$</span><span class="va">keep</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">gres</span><span class="op">$</span><span class="va">qvalue</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="co"># min nominal FDR is not 0</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 9.889241e-05</span></span></code></pre>
<div class="sourceCode" id="cb88"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">gres</span>, <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">stat</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">qvalue</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="swish_files/figure-html/unnamed-chunk-61-1.png"><!-- --></p>
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">gres</span>, <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">log2FC</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">qvalue</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v<span class="op">=</span><span class="fl">0</span>, col<span class="op">=</span><span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="swish_files/figure-html/unnamed-chunk-61-2.png"><!-- --></p>
<div class="sourceCode" id="cb90"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">gres</span>, <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">log2FC</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">qvalue</span><span class="op">)</span>,</span>
<span>                xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.5</span>,<span class="fl">1.5</span><span class="op">)</span>, ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1.5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v<span class="op">=</span><span class="fl">0</span>, col<span class="op">=</span><span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="swish_files/figure-html/unnamed-chunk-61-3.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="plotting-infrv">Plotting InfRV<a class="anchor" aria-label="anchor" href="#plotting-infrv"></a>
</h3>
<p>In the Swish paper, we describe a statistic, InfRV, which is useful
for categorizing groups of features by their inferential uncertainty.
Note that InfRV is not used in the <code>swish</code> method, but only
for visualization in the paper. Here we show how to compute and plot the
InfRV:</p>
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y3</span> <span class="op">&lt;-</span> <span class="va">se</span></span>
<span><span class="va">y3</span> <span class="op">&lt;-</span> <span class="va">y3</span><span class="op">[</span><span class="fu">seqnames</span><span class="op">(</span><span class="va">y3</span><span class="op">)</span> <span class="op">==</span> <span class="st">"chr4"</span>,<span class="op">]</span></span>
<span><span class="va">y3</span> <span class="op">&lt;-</span> <span class="va">y3</span><span class="op">[</span>,<span class="va">y3</span><span class="op">$</span><span class="va">condition</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"naive"</span>,<span class="st">"IFNg"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">y3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/labelKeep.html">labelKeep</a></span><span class="op">(</span><span class="va">y3</span><span class="op">)</span></span>
<span><span class="va">y3</span> <span class="op">&lt;-</span> <span class="va">y3</span><span class="op">[</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y3</span><span class="op">)</span><span class="op">$</span><span class="va">keep</span>,<span class="op">]</span></span>
<span><span class="va">y3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/computeInfRV.html">computeInfRV</a></span><span class="op">(</span><span class="va">y3</span><span class="op">)</span></span>
<span><span class="fu">mcols</span><span class="op">(</span><span class="va">y3</span><span class="op">)</span><span class="op">$</span><span class="va">meanCts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowMeans.html" class="external-link">rowMeans</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">y3</span><span class="op">)</span><span class="op">[[</span><span class="st">"counts"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">meanCts</span>, <span class="va">meanInfRV</span>, log<span class="op">=</span><span class="st">"xy"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="swish_files/figure-html/unnamed-chunk-62-1.png"><!-- --></p>
<div class="sourceCode" id="cb92"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">y3</span><span class="op">)</span><span class="op">$</span><span class="va">meanInfRV</span><span class="op">)</span>,</span>
<span>     col<span class="op">=</span><span class="st">"grey50"</span>, border<span class="op">=</span><span class="st">"white"</span>, breaks<span class="op">=</span><span class="fl">20</span>,</span>
<span>     xlab<span class="op">=</span><span class="st">"mean InfRV"</span>, main<span class="op">=</span><span class="st">"Txp-level inferential uncertainty"</span><span class="op">)</span></span></code></pre></div>
<p><img src="swish_files/figure-html/unnamed-chunk-62-2.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="salmon-in-alignment-mode-how-to-use-tximeta">Salmon in alignment mode, how to use tximeta<a class="anchor" aria-label="anchor" href="#salmon-in-alignment-mode-how-to-use-tximeta"></a>
</h3>
<p>If you ran Salmon in alignment mode, then a Salmon index was not
used. Without a Salmon index, there is not a checksum that tximeta uses
to identify the transcriptome. You can specify
<code>skipMeta=TRUE</code> to avoid tximeta attempting to match on the
transcriptome in this case.</p>
<p>If you want to combine transcripts to gene level (or some other level
of aggregation), you can set <code>txOut=FALSE, tx2gene=tx2gene</code>
when calling <code>tximeta</code> and it will pass these arguments to
tximport when importing the counts and inferential replicates.</p>
</div>
<div class="section level3">
<h3 id="permutation-schemes-for-interactions">Permutation schemes for interactions<a class="anchor" aria-label="anchor" href="#permutation-schemes-for-interactions"></a>
</h3>
<p>The following diagrams describe the permutation schemes used for the
interaction designs implemented in <code>swish</code>. The case with
matched samples (pair indicated by number, primary condition indicated
by color, the vertical line separating the pairs by secondary
covariate):</p>
<p><img src="swish_files/figure-html/unnamed-chunk-63-1.png"><!-- --></p>
<p>The case without matched samples (sample indicated by letter, primary
condition indicated by color, the vertical line separating the samples
by secondary covariate). Here multiple random pseudo-pairs are chosen
across condition. The permutation scheme ensures that LFCs are always
calculated between samples from the same covariate group.</p>
<p><img src="swish_files/figure-html/unnamed-chunk-64-1.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h3>
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] limma_3.57.4                GenomicFeatures_1.53.0     </span></span>
<span><span class="co">##  [3] org.Hs.eg.db_3.17.0         AnnotationDbi_1.63.1       </span></span>
<span><span class="co">##  [5] fishpond_2.7.2              tximeta_1.19.0             </span></span>
<span><span class="co">##  [7] SummarizedExperiment_1.31.1 Biobase_2.61.0             </span></span>
<span><span class="co">##  [9] GenomicRanges_1.53.1        GenomeInfoDb_1.37.1        </span></span>
<span><span class="co">## [11] IRanges_2.35.1              S4Vectors_0.39.1           </span></span>
<span><span class="co">## [13] BiocGenerics_0.47.0         MatrixGenerics_1.13.0      </span></span>
<span><span class="co">## [15] matrixStats_1.0.0           macrophage_1.17.0          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] jsonlite_1.8.5                tximport_1.29.0              </span></span>
<span><span class="co">##   [3] magrittr_2.0.3                rmarkdown_2.22               </span></span>
<span><span class="co">##   [5] fs_1.6.2                      BiocIO_1.11.0                </span></span>
<span><span class="co">##   [7] zlibbioc_1.47.0               ragg_1.2.5                   </span></span>
<span><span class="co">##   [9] vctrs_0.6.3                   memoise_2.0.1                </span></span>
<span><span class="co">##  [11] Rsamtools_2.17.0              RCurl_1.98-1.12              </span></span>
<span><span class="co">##  [13] htmltools_0.5.5               S4Arrays_1.1.4               </span></span>
<span><span class="co">##  [15] progress_1.2.2                AnnotationHub_3.9.1          </span></span>
<span><span class="co">##  [17] curl_5.0.1                    SparseArray_1.1.10           </span></span>
<span><span class="co">##  [19] sass_0.4.6                    bslib_0.5.0                  </span></span>
<span><span class="co">##  [21] desc_1.4.2                    plyr_1.8.8                   </span></span>
<span><span class="co">##  [23] cachem_1.0.8                  GenomicAlignments_1.37.0     </span></span>
<span><span class="co">##  [25] mime_0.12                     lifecycle_1.0.3              </span></span>
<span><span class="co">##  [27] pkgconfig_2.0.3               Matrix_1.5-4.1               </span></span>
<span><span class="co">##  [29] R6_2.5.1                      fastmap_1.1.1                </span></span>
<span><span class="co">##  [31] GenomeInfoDbData_1.2.10       shiny_1.7.4                  </span></span>
<span><span class="co">##  [33] digest_0.6.31                 colorspace_2.1-0             </span></span>
<span><span class="co">##  [35] rprojroot_2.0.3               textshaping_0.3.6            </span></span>
<span><span class="co">##  [37] RSQLite_2.3.1                 filelock_1.0.2               </span></span>
<span><span class="co">##  [39] fansi_1.0.4                   httr_1.4.6                   </span></span>
<span><span class="co">##  [41] abind_1.4-5                   compiler_4.3.0               </span></span>
<span><span class="co">##  [43] bit64_4.0.5                   withr_2.5.0                  </span></span>
<span><span class="co">##  [45] BiocParallel_1.35.2           DBI_1.1.3                    </span></span>
<span><span class="co">##  [47] highr_0.10                    biomaRt_2.57.1               </span></span>
<span><span class="co">##  [49] rappdirs_0.3.3                DelayedArray_0.27.5          </span></span>
<span><span class="co">##  [51] rjson_0.2.21                  gtools_3.9.4                 </span></span>
<span><span class="co">##  [53] tools_4.3.0                   interactiveDisplayBase_1.39.0</span></span>
<span><span class="co">##  [55] httpuv_1.6.11                 glue_1.6.2                   </span></span>
<span><span class="co">##  [57] restfulr_0.0.15               promises_1.2.0.1             </span></span>
<span><span class="co">##  [59] grid_4.3.0                    reshape2_1.4.4               </span></span>
<span><span class="co">##  [61] generics_0.1.3                gtable_0.3.3                 </span></span>
<span><span class="co">##  [63] ensembldb_2.25.0              hms_1.1.3                    </span></span>
<span><span class="co">##  [65] xml2_1.3.4                    utf8_1.2.3                   </span></span>
<span><span class="co">##  [67] XVector_0.41.1                BiocVersion_3.18.0           </span></span>
<span><span class="co">##  [69] pillar_1.9.0                  stringr_1.5.0                </span></span>
<span><span class="co">##  [71] later_1.3.1                   splines_4.3.0                </span></span>
<span><span class="co">##  [73] dplyr_1.1.2                   BiocFileCache_2.9.0          </span></span>
<span><span class="co">##  [75] lattice_0.21-8                rtracklayer_1.61.0           </span></span>
<span><span class="co">##  [77] bit_4.0.5                     tidyselect_1.2.0             </span></span>
<span><span class="co">##  [79] SingleCellExperiment_1.23.0   Biostrings_2.69.1            </span></span>
<span><span class="co">##  [81] knitr_1.43                    ProtGenerics_1.33.1          </span></span>
<span><span class="co">##  [83] xfun_0.39                     stringi_1.7.12               </span></span>
<span><span class="co">##  [85] lazyeval_0.2.2                yaml_2.3.7                   </span></span>
<span><span class="co">##  [87] evaluate_0.21                 codetools_0.2-19             </span></span>
<span><span class="co">##  [89] tibble_3.2.1                  qvalue_2.33.0                </span></span>
<span><span class="co">##  [91] BiocManager_1.30.21           cli_3.6.1                    </span></span>
<span><span class="co">##  [93] xtable_1.8-4                  systemfonts_1.0.4            </span></span>
<span><span class="co">##  [95] munsell_0.5.0                 jquerylib_0.1.4              </span></span>
<span><span class="co">##  [97] Rcpp_1.0.10                   svMisc_1.2.3                 </span></span>
<span><span class="co">##  [99] dbplyr_2.3.2                  png_0.1-8                    </span></span>
<span><span class="co">## [101] XML_3.99-0.14                 parallel_4.3.0               </span></span>
<span><span class="co">## [103] ellipsis_0.3.2                pkgdown_2.0.7                </span></span>
<span><span class="co">## [105] ggplot2_3.4.2                 blob_1.2.4                   </span></span>
<span><span class="co">## [107] prettyunits_1.1.1             AnnotationFilter_1.25.0      </span></span>
<span><span class="co">## [109] bitops_1.0-7                  scales_1.2.1                 </span></span>
<span><span class="co">## [111] purrr_1.0.1                   crayon_1.5.2                 </span></span>
<span><span class="co">## [113] rlang_1.1.1                   KEGGREST_1.41.0</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-alasoo" class="csl-entry">
Alasoo, K, J Rodrigues, S Mukhopadhyay, AJ Knights, AL Mann, K Kundu,
HIPSCI-Consortium, C Hale, Dougan G, and DJ Gaffney. 2018. <span>“<span class="nocase">Shared genetic effects on chromatin and gene expression
indicate a role for enhancer priming in immune response</span>.”</span>
<em>Nature Genetics</em> 50: 424–31. <a href="https://doi.org/10.1038/s41588-018-0046-7" class="external-link">https://doi.org/10.1038/s41588-018-0046-7</a>.
</div>
<div id="ref-Amezquita2020" class="csl-entry">
Amezquita, Robert A., Aaron T. L. Lun, Etienne Becht, Vince J. Carey,
Lindsay N. Carpp, Ludwig Geistlinger, Federico Marini, et al. 2020.
<span>“Orchestrating Single-Cell Analysis with Bioconductor.”</span>
<em>Nature Methods</em> 17 (2): 137–45. <a href="https://doi.org/10.1038/s41592-019-0654-x" class="external-link">https://doi.org/10.1038/s41592-019-0654-x</a>.
</div>
<div id="ref-anderson" class="csl-entry">
Anderson, MJ, and CJF Ter Braak. 2003. <span>“Permutation Tests for
Multi-Factorial Analysis of Variance.”</span> <em>Journal of Statistical
Computation and Simulation</em> 73 (2): 85–113.
</div>
<div id="ref-gencode" class="csl-entry">
Frankish, A, GENCODE-consoritum, and P Flicek. 2018. <span>“<span class="nocase">GENCODE reference annotation for the human and mouse
genomes</span>.”</span> <em>Nucleic Acids Research</em>. <a href="https://doi.org/10.1093/nar/gky955" class="external-link">https://doi.org/10.1093/nar/gky955</a>.
</div>
<div id="ref-samseq" class="csl-entry">
Li, J, and R Tibshirani. 2011. <span>“<span class="nocase">Finding
consistent patterns: A nonparametric approach for identifying
differential expression in RNA-Seq data</span>.”</span> <em>Statistical
Methods in Medical Research</em> 22 (5): 519–36. <a href="https://doi.org/10.1177/0962280211428386" class="external-link">https://doi.org/10.1177/0962280211428386</a>.
</div>
<div id="ref-tximeta" class="csl-entry">
Love, Michael I., Charlotte Soneson, Peter F. Hickey, Lisa K. Johnson,
N. Tessa Pierce, Lori Shepherd, Martin Morgan, and Rob Patro. 2020.
<span>“<span class="nocase">Tximeta: Reference sequence checksums for
provenance identification in RNA-seq</span>.”</span> <em>PLOS
Computational Biology</em>. <a href="https://doi.org/10.1371/journal.pcbi.1007664" class="external-link">https://doi.org/10.1371/journal.pcbi.1007664</a>.
</div>
<div id="ref-scran" class="csl-entry">
Lun, A. T. L., D. J. McCarthy, and J. C. Marioni. 2016. <span>“A
Step-by-Step Workflow for Low-Level Analysis of Single-Cell RNA-Seq Data
with Bioconductor.”</span> <em>F1000Research</em> 5: 2122. <a href="https://doi.org/10.12688/f1000research.9501.2" class="external-link">https://doi.org/10.12688/f1000research.9501.2</a>.
</div>
<div id="ref-Lun2016" class="csl-entry">
Lun, Aaron T. L., Karsten Bach, and John C. Marioni. 2016. <span>“<span class="nocase">Pooling across cells to normalize single-cell RNA
sequencing data with many zero counts</span>.”</span> <em>Genome
Biology</em> 17 (1): 75. <a href="https://doi.org/10.1186/s13059-016-0947-7" class="external-link">https://doi.org/10.1186/s13059-016-0947-7</a>.
</div>
<div id="ref-salmon" class="csl-entry">
Patro, R, G Duggal, MI Love, RA Irizarry, and C Kingsford. 2017.
<span>“Salmon Provides Fast and Bias-Aware Quantification of Transcript
Expression.”</span> <em>Nature Methods</em> 14: 417–19. <a href="https://doi.org/10.1038/nmeth.4197" class="external-link">https://doi.org/10.1038/nmeth.4197</a>.
</div>
<div id="ref-alevin" class="csl-entry">
Srivastava, Avi, Laraib Malik, Tom Sean Smith, Ian Sudbery, and Rob
Patro. 2019. <span>“<span class="nocase">Alevin efficiently estimates
accurate gene abundances from dscRNA-seq data</span>.”</span> <em>Genome
Biology</em> 20 (65). <a href="https://doi.org/10.1186/s13059-019-1670-y" class="external-link">https://doi.org/10.1186/s13059-019-1670-y</a>.
</div>
<div id="ref-qvalue" class="csl-entry">
Storey, JD, and R Tibshirani. 2003.<span>“<span class="nocase">
Statistical significance for genome-wide experiments</span>.”</span>
<em>Proceedings of the National Academy of Sciences</em> 100 (16):
9440–45. <a href="https://doi.org/10.1073/pnas.1530509100" class="external-link">https://doi.org/10.1073/pnas.1530509100</a>.
</div>
<div id="ref-compression" class="csl-entry">
Van Buren, S, H Sarkar, A Srivastava, NU Rashid, R Patro, and MI Love.
2020. <span>“<span class="nocase">Compression of quantification
uncertainty for scRNA-seq counts</span>.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2020.07.06.189639" class="external-link">https://doi.org/10.1101/2020.07.06.189639</a>.
</div>
<div id="ref-wu2022" class="csl-entry">
Wu, Euphy, Noor P. Singh, Kwangbom Choi, Mohsen Zakeri, Matthew Vincent,
Gary A. Churchill, Cheryl L. Ackert-Bicknell, Rob Patro, and Michael I.
Love. 2022. <span>“Detecting Isoform-Level Allelic Imbalance Accounting
for Inferential Uncertainty.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2022.08.12.503785" class="external-link">https://doi.org/10.1101/2022.08.12.503785</a>.
</div>
<div id="ref-swish" class="csl-entry">
Zhu, A, A Srivastava, JG Ibrahim, R Patro, and MI Love. 2019.
<span>“<span class="nocase">Nonparametric expression analysis using
inferential replicate counts</span>.”</span> <em>Nucleic Acids
Research</em> 47: e105. <a href="https://doi.org/10.1093/nar/gkz622" class="external-link">https://doi.org/10.1093/nar/gkz622</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://www.linkedin.com/in/aqzhu91" class="external-link">Anqi Zhu</a>, <a href="https://mikelove.github.io" class="external-link">Michael Love</a>, <a href="https://k3yavi.github.io" class="external-link">Avi Srivastava</a>, <a href="https://combine-lab.github.io" class="external-link">Rob Patro</a>, <a href="https://sph.unc.edu/adv_profile/joseph-g-ibrahim-phd" class="external-link">Joseph Ibrahim</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
